---
title: "LCM_ATAC_revision_merged"
author: "C Carraro"
date: "14/07/23"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Install packages

```{r}
list.of.bioc.packages<- c("Rsamtools",
                          "GenomicAlignments",
                          "rtracklayer",
                          "ChIPseeker",
                          "DBI",
                          "TxDb.Mmusculus.UCSC.mm10.knownGene",
                          "org.Mm.eg.db",
                          "DESeq2",
                          "org.Hs.eg.db",
                          "vsn",
                          "tximport",
                          "rhdf5",
                          "clusterProfiler",
                          "DOSE",
                          "GSEABase",
                          "RColorBrewer",
                          "ComplexHeatmap",
                          "pheatmap",
                          "genefilter",
                          "biomaRt",
                          "limma",
                          "sva",
                          "IHW",
                          "Gviz",
                          "tidyverse",
                          "useful",
                          "gplots",
                          "SummarizedExperiment",
                          "BiocParallel",
                          "ggrepel",
                          "ggbeeswarm",
                          "grid",
                          "gridextra",
                          "reshape2",
                          "factoextra",
                          "Hmisc",
                          "GenomicRanges",
                          "BSgenome.Mmusculus.UCSC.mm10",
                          "cowplot",
                          "HelloRanges",
                          "ATACseqQC",
                          "BRGenomics",
                          "UpSetR",
                          "GSVA")
new.packages.bioc <- list.of.bioc.packages[!(list.of.bioc.packages %in% installed.packages()[,"Package"])]
 
if(length(new.packages.bioc)>0)if (!requireNamespace("BiocManager")) install.packages("BiocManager")
BiocManager::install(new.packages.bioc, update = FALSE)
```

### Load packages

```{r, message=FALSE}
library(circlize)
library(viridis)
library(viridisLite)
library(tidyverse)
library(useful)
library(gplots)
library(SummarizedExperiment)
library(BiocParallel)
library(ggrepel)
library(grid)
library(gridExtra)
library(reshape2)
library(factoextra)
library(Hmisc)
library(GenomicRanges)
library(GenomicAlignments)
library(rtracklayer)
library(Rsamtools)
library(ChIPseeker)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(biomaRt)
library(org.Mm.eg.db) 
library(org.Hs.eg.db) 
library(DESeq2)
library(tximport)
library(vsn)
library(IHW)
library(rhdf5)
library(genefilter)
library(fdrtool)
library(RColorBrewer)
library(pheatmap)
library(cowplot) 
library(Gviz)
library(GSEABase)
library(clusterProfiler)
library(DOSE)
library(sva)
library(limma)
library(BSgenome.Mmusculus.UCSC.mm10)
library(UpSetR)
library(HelloRanges)
library(ComplexHeatmap)
library(ATACseqQC)
library(BRGenomics)
library(GSVA)
library(fgsea)
library(ggpubr)
library(chromVAR)
library(motifmatchr)
```

# ATAC-seq analysis

## Functions

### Peak import

```{r}
# Importing of BED files and production of GRanges object of all peaks
# Modified to support multiple files exclusion
importBEDpeaks <- function(pattern = NULL, 
                           path = NULL,
                           exclude = NULL){
  fileNames <- dir(path, pattern = pattern)
  if(!is.null(exclude)){
    for (i in exclude){
  fileNames <- fileNames[!grepl(i, fileNames)] 
  }
  }
  bedPeaks <- GRangesList()
  for (i in fileNames) {
    tmp <- read.table(paste(path,i,sep="/"))
    bedPeaks[[i]] <- GRanges(seqnames = Rle(tmp$V1), ranges = IRanges(start = tmp$V2, end = tmp$V3), strand = Rle("*"))
    bedPeaks[[i]] <- bedPeaks[[i]][grepl("chr",bedPeaks[[i]]@seqnames),]
  }
  if (!is.null(pattern)) names(bedPeaks) <- sub(pattern, "", names(bedPeaks), fixed = TRUE)
  return(bedPeaks)
}
```

### Peak unification

```{r}
# Produce union of all peaks for counting of the reads
unionPeaks <- function(peaks) {
  upks <- GenomicRanges::reduce(unlist(peaks))
  for (i in names(peaks)) {
    mcols(upks)[, i] <- countOverlaps(subject = peaks[[i]],
                                      query = upks) != 0
  }
  names(upks) <- paste(seqnames(upks), start(upks), end(upks), sep="_")
  return(upks)
}
```

### Count table

```{r}
# Prepare count table for all peaks
peakCountTable <- function(peaks, path = NULL, pattern = NULL,
                           exclude = NULL) {
  fileNames <- dir(path, pattern = pattern)
  if(!is.null(exclude)){
    for (i in exclude){
  fileNames <- fileNames[!grepl(i, fileNames)] 
  }
  }
  bamFiles <- BamFileList(paste(path,fileNames,sep="/"), asMates = TRUE)
  countTable <- summarizeOverlaps(features = peaks,
                                  reads = bamFiles,
                                  mode = "Union",
                                  singleEnd = FALSE,
                                  fragments =FALSE,
                                  ignore.strand = TRUE)
  if (!is.null(pattern)) colnames(countTable) <- sub(pattern, "", colnames(countTable))
  return(countTable)
}
```

### PCA

```{r}
plotPCA <- function(pca_input = dds_vst,
                    ntop="all", 
                    xPC=1, 
                    yPC=2,
                    point_size=3,
                    title="",
                    color,
                    #shape,
                    anno_color="NULL"){
  
  vst_matrix <- as.matrix(assay(pca_input))
  
  if(ntop=="all"){
    pca <- prcomp(t(vst_matrix)) 
  }else{
    # select the ntop genes by variance
    select <- order(rowVars(vst_matrix), decreasing=TRUE)[c(1:ntop)]
    pca <- prcomp(t(vst_matrix[select,]))
  }
  
  #calculate explained variance per PC
  explVar <- pca$sdev^2/sum(pca$sdev^2)
  # transform variance to percent
  percentVar <- round(100 * explVar[c(xPC,yPC)], digits=1)
  
  # Define data for plotting  
  pcaData <- data.frame(xPC=pca$x[,xPC], 
                        yPC=pca$x[,yPC], 
                        color = sample_table[[color]],
                        #shape = sample_table[[shape]],
                        name= as.character(sample_table$ID),
                        stringsAsFactors = F)
  
  #plot PCA
  pca_plot <- ggplot(pcaData, aes(x = xPC, y = yPC, fill=color)) +
    geom_point(size =point_size, shape=21, color="black") + theme_bw()+
      xlab(paste0("PC ",xPC, ": ", percentVar[1], "% variance")) +
    ylab(paste0("PC ",yPC,": ", percentVar[2], "% variance"))
  
  if(anno_color[1] == "NULL"){
    pca_plot <- pca_plot + scale_color_discrete(name=color)
  }else{
    pca_plot <- pca_plot + scale_fill_manual(values=anno_color)
  }
  
  # pca_plot <- pca_plot +
  # geom_text_repel(aes(label = name), colour = "black")+
  # xlab(paste0("PC ",xPC, ": ", percentVar[1], "% variance")) +
  # ylab(paste0("PC ",yPC,": ", percentVar[2], "% variance")) +
  # coord_fixed()+
  # theme_classic()+
  # theme(aspect.ratio = 1)+
  # ggtitle(title)
  # 
  # pca_plot
}
```

### DESeq2 output

```{r}
# Specify structure of DESeq2_analysis_object

setClass(Class = "DESeq2_analysis_object",
         slots = c(results="data.frame", DE_genes="list", Number_DE_genes="list"))


# Wrapper Function to perform DESeq2 differential testing
DEAnalysis <- function(condition,
                       alpha = 0.05, 
                       lfcThreshold = 0,
                       sigFC = 2, 
                       multiple_testing = "IHW",
                       independentFiltering="TRUE",
                       shrinkage = TRUE,
                       shrinkType = "normal"){
  # create results_list  
  results_list <- list()
  # print parameters
  results_list$parameters <-list(multiple_testing = multiple_testing,
                                 p_value_threshold = alpha,
                                 log2_FC_threshold = lfcThreshold,
                                 shrinkage = shrinkage,
                                 shrinkage_type = shrinkType)
  # Run results() function on comparisons defined in comparison table
  for (i in 1:nrow(comparison_table)){
    # create DE_object
    DE_object <- new(Class = "DESeq2_analysis_object")
    # IHW
    if (multiple_testing=="IHW") {
      res_deseq_lfc <- results(dds,
                               contrast = c(condition,
                                            paste(comparison_table$comparison[i]),
                                            paste(comparison_table$control[i])),
                               lfcThreshold = lfcThreshold,
                               alpha = alpha,
                               filterFun = ihw,
                               altHypothesis = "greaterAbs")
      # Independent Filtering
    }else {
      res_deseq_lfc <- results(dds,
                               contrast = c(condition,
                                            paste(comparison_table$comparison[i]),
                                            paste(comparison_table$control[i])),
                               lfcThreshold = lfcThreshold,
                               alpha = alpha,
                               independentFiltering = independentFiltering,
                               altHypothesis = "greaterAbs",
                               pAdjustMethod= multiple_testing)
    }
    if(shrinkage == TRUE){
      res_deseq_lfc <- lfcShrink(dds, 
                                 contrast = c(condition,
                                              paste(comparison_table$comparison[i]),
                                              paste(comparison_table$control[i])),
                                 res=res_deseq_lfc,
                                 type = shrinkType)
    }
    res_deseq_lfc <- as.data.frame(res_deseq_lfc)
    # indicate significant DE genes  
    res_deseq_lfc$regulation <- ifelse(!is.na(res_deseq_lfc$padj)&
                                         res_deseq_lfc$padj <= alpha&
                                         res_deseq_lfc$log2FoldChange > log(sigFC,2),
                                       "up",
                                       ifelse(!is.na(res_deseq_lfc$padj)&
                                                res_deseq_lfc$padj <= alpha&
                                                res_deseq_lfc$log2FoldChange < -log(sigFC,2),
                                              "down",
                                              "n.s."))
    # add gene annotation to results table
    res_deseq_lfc$peakid <- row.names(res_deseq_lfc) # ensembl-IDs as row names
    res_deseq_lfc <- merge(res_deseq_lfc, 
                           norm_anno[,c("peakid", 
                                        "symbol",
                                        "geneChr",
                                        "geneStart",
                                        "geneEnd",
                                        "geneLength",
                                        "geneStrand",
                                        "geneId",
                                        "distanceToTSS",
                                        "start",
                                        "end",
                                        "annotation")], 
                           by = "peakid") 
    row.names(res_deseq_lfc) <- res_deseq_lfc$peakid
    res_deseq_lfc$comparison<-paste(comparison_table$comparison[i]," vs ",comparison_table$control[i],
                                    sep="")
    # re-order results table
    if (multiple_testing=="IHW") {
      res_deseq_lfc<-res_deseq_lfc[,c("peakid", 
                                        "symbol",
                                      "geneId",
                                        "geneChr",
                                     "comparison",
                                     "regulation",
                                     "baseMean",
                                     "log2FoldChange",
                                     "lfcSE",
                                     "stat",
                                     "pvalue",
                                     "padj",
                                     "weight",
                                     "annotation")]
    }else{
      res_deseq_lfc<-res_deseq_lfc[,c("peakid", 
                                        "symbol",
                                      "geneId",
                                        "geneChr",
                                      "comparison",
                                      "regulation",
                                      "baseMean",
                                      "log2FoldChange",
                                      "lfcSE",
                                      "stat",
                                      "pvalue",
                                      "padj",
                                      "annotation")]
    }
    # print result table
    DE_object@results <- res_deseq_lfc
    # print DE genes in seperate tables
    DE_object@DE_genes <- list(up_regulated_Genes = res_deseq_lfc[res_deseq_lfc$regulation =="up",],
                               down_regulated_Genes= res_deseq_lfc[res_deseq_lfc$regulation =="down",])
    # print the numbers of DE genes
    DE_object@Number_DE_genes <- list(up_regulated_Genes = nrow(DE_object@DE_genes$up_regulated_Genes),
                                      down_regulated_Genes= nrow(DE_object@DE_genes$down_regulated_Genes))
    # write DE_object into results_list
    results_list[[paste(comparison_table$comparison[i], "vs", comparison_table$control[i], sep="_")]] <- DE_object
  }
  return(results_list)
}
```

### Union of DAR (all)

```{r}
uDARall <- function(comparisons){
  uDARs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DARs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation %in% c("up","down"),])
    uDARs <- unique(c(uDARs, DARs$peakid))
  }
  uDARs
}
```

### Union of DAR (up)

```{r}
uDARall_up <- function(comparisons){
  uDARs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DARs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation == "up",])
    uDARs <- unique(c(uDARs, DARs$peakid))
  }
  uDARs
}
```

### Union of DAR (down)

```{r}
uDARall_down <- function(comparisons){
  uDARs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DARs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation == "down",])
    uDARs <- unique(c(uDARs, DARs$peakid))
  }
  uDARs
}
```

### Union of DAR (promoter regions)

```{r}
uDAR_promo <- function(comparisons){
  uDARs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DARs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation %in% c("up","down"),])
    DARs <- subset(DARs, DARs$annotation %in% c("Promoter (<=1kb)", "Promoter (1-2kb)", "Promoter (2-3kb)"))
    uDARs <- unique(c(uDARs, DARs$peakid))
  }
  uDARs
}
```

### Union of DAR up (promoter regions)

```{r}
uDAR_promo_up <- function(comparisons){
  uDARs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DARs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation == "up",])
    DARs <- subset(DARs, DARs$annotation %in% c("Promoter (<=1kb)", "Promoter (1-2kb)", "Promoter (2-3kb)"))
    uDARs <- unique(c(uDARs, DARs$peakid))
  }
  uDARs
}
```

### Union of DAR down (promoter regions)

```{r}
uDAR_promo_down <- function(comparisons){
  uDARs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DARs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation == "down",])
    DARs <- subset(DARs, DARs$annotation %in% c("Promoter (<=1kb)", "Promoter (1-2kb)", "Promoter (2-3kb)"))
    uDARs <- unique(c(uDARs, DARs$peakid))
  }
  uDARs
}
```

### Union of DEgenes (promoter regions)

```{r}
uDEG <- function(comparisons){
  uDEGs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DEGs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation %in% c("up","down"),])
    DEGs <- subset(DEGs, DEGs$annotation %in% c("Promoter (<=1kb)", "Promoter (1-2kb)", "Promoter (2-3kb)"))
    uDEGs <- unique(c(uDEGs, DEGs$symbol))
  }
  uDEGs
}
```

### Union of DEgenes UP (promoter regions)

```{r}
uDEGup <- function(comparisons){
  uDEGs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DEGs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation %in% c("up"),])
  DEGs <- subset(DEGs, DEGs$annotation %in% c("Promoter (<=1kb)", "Promoter (1-2kb)", "Promoter (2-3kb)"))
    uDEGs <- unique(c(uDEGs, DEGs$symbol))
  }
  uDEGs
}
```

### Union of DEgenes DOWN (promoter regions)

```{r}
uDEGdown <- function(comparisons){
  uDEGs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DEGs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation %in% c("down"),])
    DEGs <- subset(DEGs, DEGs$annotation %in% c("Promoter (<=1kb)", "Promoter (1-2kb)", "Promoter (2-3kb)"))
    uDEGs <- unique(c(uDEGs, DEGs$symbol))
  }
  uDEGs
}
```

###GSEA dotplot for multiple comparisons

```{r}
dotplotGSEA_MC <- function(x,
                        show=25,
                        font.size=10,
                        title.size=10,
                        title.width=100){
    ggplot(x, aes(x = Cluster, y = Description, color = qvalue)) +
      geom_point(aes(size = Count)) +
      scale_x_discrete(drop = FALSE) +
      scale_colour_gradientn(colours=c('darkorange1',
                                       '#FFFF0000',
                                       '#FFFF0000'),
                             limits=c(0,1),
                             values   = c(0,0.25,0.26,1),
                             breaks   = c(0.25,0.26,1),
                             labels = format(c(0.25,0.26,1))) +
      ylab(NULL) +
      # ggtitle(paste(strwrap(unique(x$Enrichment), width=title.width), collapse = "\n"))+
      theme_bw() +
      theme(text = element_text(size=font.size),
            plot.title = element_text(size=title.size), axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
  }
```

### GO & KEGG enrichment across comparisons

```{r}
compareGSEA <- function(comparisons, 
                        organism, # chose organism
                        GeneSets =c("GO","KEGG"), # choose gene sets for enrichment
                        ontology= "BP", # define GO subset
                        pCorrection = "bonferroni", # choose the p-value adjustment method
                        pvalueCutoff = 0.05, # set the unadj. or adj. p-value cutoff (depending on correction method)
                        qvalueCutoff = 0.05, # set the q-value cutoff (FDR corrected)
                        showMax = 20){
  
  if(organism == "mouse") {
    OrgDb = org.Mm.eg.db
  } else if(organism == "human"){
      OrgDb = org.Hs.eg.db
    } else {stop("Wrong Organism. Select mouse or human.")}
  
  ENTREZlist <-  list()
  for(i in 1:length(comparisons)){
    res <- DEresults[names(DEresults) %in% comparisons]
    DE_up <- as.data.frame(res[[i]]@DE_genes$up_regulated_Genes)$symbol
    entrez_up <- bitr(DE_up, fromType = "SYMBOL", toType="ENTREZID", OrgDb=OrgDb)$ENTREZID
    DE_down <- as.data.frame(res[[i]]@DE_genes$down_regulated_Genes)$symbol
    entrez_down <- bitr(DE_down, fromType = "SYMBOL", toType="ENTREZID", OrgDb=OrgDb)$ENTREZID  
    x <- setNames(list(entrez_up, entrez_down),
                  c(paste(names(res[i]),"_up",sep=""), 
                    paste(names(res[i]),"_down",sep="")))
    ENTREZlist <- c(ENTREZlist,x)
  }
  
  list <- list()
  
  # Compare the Clusters regarding their GO enrichment  
  if("GO" %in% GeneSets){
    print("Performing GO enrichment")
    CompareClusters_GO <- compareCluster(geneCluster = ENTREZlist, 
                                       fun = "enrichGO",  
                                       universe = universe_Entrez,
                                       OrgDb = OrgDb,
                                       ont = ontology, 
                                       pvalueCutoff  = pvalueCutoff, 
                                       pAdjustMethod = pCorrection, 
                                       qvalueCutoff  = pvalueCutoff,  
                                       readable      = T)
    list$GOresults <- as.data.frame(CompareClusters_GO)
    list$GOplot <- clusterProfiler::dotplot(CompareClusters_GO, showCategory = showMax, by = "geneRatio", font.size=10)
  }
  
  if("KEGG" %in% GeneSets){
    print("Performing KEGG enrichment")
    
    if(organism == "mouse"){org = "mmu"} 
    if(organism == "human"){org = "hsa"}
    
    # Compare the Clusters regarding their KEGG enrichment  
    CompareClusters_KEGG <- compareCluster(geneCluster = ENTREZlist, 
                                         fun = "enrichKEGG",  
                                         universe = universe_Entrez,
                                         organism = org, 
                                         pvalueCutoff  = pvalueCutoff, 
                                         pAdjustMethod = pCorrection, 
                                         qvalueCutoff  = pvalueCutoff)
    list$KEGGresults <- as.data.frame(CompareClusters_KEGG)
    list$KEGGplot <- clusterProfiler::dotplot(CompareClusters_KEGG, showCategory = showMax, by = "geneRatio", font.size=10)
  }
  list
}
```

## Directories

```{r}
projectPath <- "/home/caterina/data/alignment/2023-07-18_selectedsamples/"
bamFilePath <- paste(projectPath,"output/mapped", sep = "")
bedFilePath <- paste(projectPath, "output/peaks", sep = "")

analysisPath <- "/home/caterina/data/analysis/"
```

## Sample Annotation 

```{r}
sample_table <- read.csv("/home/caterina/data/analysis/sampletable_lcm0723_updated.csv",sep=";",header=TRUE, stringsAsFactors = TRUE)

sample_table$ID <- sample_table$PRECISE_ID
sample_table$ID <- paste("Sample_",sample_table$PRECISE_ID,sep="")
rownames(sample_table) <- sample_table$PRECISE_ID
```

### Colour scheme

For hex colors use: https://www.color-hex.com

```{r}
# Tissue

col_CELL_TYPE <- c("#f25814", "#0353aa")

names(col_CELL_TYPE) <- c("Bcell", "Tcell") 

# Condition

col_GROUP = c("#ce7e00", "#ffd966", "#0e2f44", "#468499")

names(col_GROUP) <- c("B_patch", "B_100", "T_patch", "T_100")

# Cell_number

col_CUT_TYPE = c("#feda75", "#fa7e1e")

names(col_CUT_TYPE) <- c("patch", "100cells")

# combine color code into list

anno_color <- list(GROUP = col_GROUP, 
                  CELL_TYPE = col_CELL_TYPE, 
                  CUT_TYPE = col_CUT_TYPE)
```

## Data Import, QC and Processing

### TSS enrichment score

```{r}
set.seed(1)

txs <- transcripts(TxDb.Mmusculus.UCSC.mm10.knownGene)
TSS_df <- list()

for(i in 1:length(sample_table$ID)){

  message(dir(path=bamFilePath, pattern = ".final.bam$")[i])
  sortedBAM <- dir(path=bamFilePath, pattern = ".final.bam$",full.names = TRUE)[i]
  sample <- paste(strsplit(dir(path=bamFilePath, pattern = ".final.bam$")[i],"\\.")[[1]][1],sep="")

  bam <- import_bam(sortedBAM,
                  mapq = 20, #Filter reads by a minimum MAPQ score. This is the correct way to filter multi-aligners.
                  revcomp = FALSE, #Logical indicating if aligned reads should be reverse-complemented.
                  shift = 0L,
                  trim.to = c("whole", "5p", "3p", "center"), #Option for selecting specific bases from the reads, applied after the revcomp and shift options. By default, the entire read is maintained.
                  ignore.strand = FALSE, #Logical indicating if the strand information should be discarded.
                  field = "score", #Metadata field name to use for readcounts, usually "score". If set to NULL, identical reads (or identical positions if trim.to options applied) are not combined, and the length of the output GRanges will be equal to the number of input reads.
                  paired_end = TRUE)

  bam_gal <- as(bam, "GAlignments")

  tsse <- TSSEscore(bam_gal, txs)

  TSS_df[[paste0(sample)]] <- tsse$TSSEscore

}
```

### Insert size

```{r,fig.height=24,fig.width=12}
plotList_2 <- list()

for(i in 1:length(sample_table$ID)){
  message(dir(path=bamFilePath, pattern = ".final.bam$")[i])
  sortedBAM <- dir(path=bamFilePath, pattern = ".final.bam$",full.names = TRUE)[i]
  sample <- paste(strsplit(dir(path=bamFilePath, pattern = ".final.bam$")[i],"\\.")[[1]][1],sep="")

  atacReads <- readGAlignmentPairs(sortedBAM,
                                   param = ScanBamParam(mapqFilter = 1,
                                                        flag = scanBamFlag(isPaired = TRUE, isProperPair = TRUE),
                                                        what = c("qname",
                                                                 "mapq",
                                                                 "isize")))

  atacReads_read1 <- GenomicAlignments::first(atacReads)
  insertSizes <- abs(elementMetadata(atacReads_read1)$isize)

  inserts <- table(insertSizes) %>%
    data.frame %>%
    mutate(insertSizes = as.numeric(as.vector(insertSizes)),
           Freq = as.numeric(as.vector(Freq)))

  plotList_2[[i]] <- ggplot(inserts, aes(x = insertSizes, y = Freq)) +
    geom_line()+
    geom_vline(xintercept = c(180, 247), colour = "red") +
    geom_vline(xintercept = c(315, 437), colour = "darkblue") +
    geom_vline(xintercept = c(100), colour = "darkgreen") +
    ggtitle(sample)+
    theme_bw()
}

plot_grid(plotlist = plotList_2,ncol=2)
```

#### Overlayed FSD

```{r, fig.width=5, fig.height=4}
inserts <- list()

for(i in 1:length(plotList_2)){

  m <- plotList_2[[i]][["data"]]
  q <- plotList_2[[i]][["labels"]][["title"]]
  inserts[[paste0(q)]] <- m

}

inserts_df <- data.frame()

for(i in names(inserts)){

  n <- inserts[[i]]
  n$df <- paste0(i)

  inserts_df <- rbind(inserts_df, n)

}

levels <- sample_table$ID
inserts_df$df <- factor(inserts_df$df, levels = levels)

fsd_n <- ggplot(inserts_df, aes(x = insertSizes, y = Freq)) +
  geom_line(aes(color = df)) +
  xlim(0, 800) +
  ylim(0, 300000) +
  labs(x="Fragment size (bp)", y="Fragments number") +
  theme_classic() +
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10)) +
  theme(aspect.ratio = 1)

insert_subset <- inserts_df
insert_subset <- insert_subset %>% group_by(df) %>% mutate(sum = sum(Freq))
insert_subset$ratio <- ((insert_subset$Freq)*100)/(insert_subset$sum)

fsd_ratio <- ggplot(insert_subset, aes(x = insertSizes, y = ratio)) +
  geom_line(aes(color = df)) +
  xlim(0, 800) +
  ylim(0, 1.5) +
  labs(x="Fragment size (bp)", y="Fragments %") +
  theme_classic() +
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10)) +
  theme(aspect.ratio = 1)

fsd_n
fsd_ratio
```

#### Plot groupwise

```{r}
insert_subset_Bpatch <- subset(insert_subset, insert_subset$df %in% c("Sample_24153", "Sample_24155", "Sample_24158", "Sample_24160", "Sample_24162"))

insert_subset_Tpatch <- subset(insert_subset, insert_subset$df %in% c("Sample_24154", "Sample_24156", "Sample_24157", "Sample_24159", "Sample_24161"))

insert_subset_B100 <- subset(insert_subset, insert_subset$df %in% c("Sample_24163", "Sample_24164", "Sample_24165", "Sample_24167"))

insert_subset_T100 <- subset(insert_subset, insert_subset$df %in% c("Sample_24168", "Sample_24169", "Sample_24170", "Sample_24171"))

fsd_Bpatch <- ggplot(insert_subset_Bpatch, aes(x = insertSizes, y = ratio)) +
  geom_line(aes(color = df)) +
  scale_colour_manual(values=c("#f0d8b2", "#e1b166", "#ce7e00", "#905800", "#673f00")) +
  xlim(0, 800) +
  ylim(0, 1.5) +
  labs(x="Fragment size (bp)", y="Fragments %") +
  theme_classic() +
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10)) +
  theme(aspect.ratio = 1) +
  ggtitle("Bpatch")

fsd_Tpatch <- ggplot(insert_subset_Tpatch, aes(x = insertSizes, y = ratio)) +
  geom_line(aes(color = df)) +
  scale_colour_manual(values=c("#ced5d9", "#9eabb4", "#6e828e", "#3e5869", "#0e2f44")) +
  xlim(0, 800) +
  ylim(0, 1.5) +
  labs(x="Fragment size (bp)", y="Fragments %") +
  theme_classic() +
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10)) +
  theme(aspect.ratio = 1) +
  ggtitle("Tpatch")

fsd_B100 <- ggplot(insert_subset_B100, aes(x = insertSizes, y = ratio)) +
  geom_line(aes(color = df)) +
  scale_colour_manual(values=c("#f0d8b2", "#ce7e00", "#905800", "#673f00")) +
  xlim(0, 800) +
  ylim(0, 1.5) +
  labs(x="Fragment size (bp)", y="Fragments %") +
  theme_classic() +
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10)) +
  theme(aspect.ratio = 1) +
  ggtitle("B100")

fsd_T100 <- ggplot(insert_subset_T100, aes(x = insertSizes, y = ratio)) +
  geom_line(aes(color = df)) +
  scale_colour_manual(values=c("#ced5d9", "#9eabb4", "#6e828e", "#3e5869")) +
  xlim(0, 800) +
  ylim(0, 1.5) +
  labs(x="Fragment size (bp)", y="Fragments %") +
  theme_classic() +
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10)) +
  theme(aspect.ratio = 1) +
  ggtitle("T100")

grid.arrange(fsd_Bpatch, fsd_Tpatch, fsd_B100, fsd_T100, ncol=2, nrow=2)
```

### Import merged bam-called peaks

Peaks are here called on bed derived from merging all bam files (lung and brain samples). 

```{r}
tmp <- read.table(paste(bedFilePath, "merged_peaks.narrowPeak", sep="/"))
peakSummary <- GRanges(seqnames = Rle(tmp$V1), ranges = IRanges(start = tmp$V2, end = tmp$V3), strand = Rle("*"))
peakSummary <- peakSummary[grepl("chr",peakSummary@seqnames),]
names(peakSummary) <- paste(seqnames(peakSummary), start(peakSummary), end(peakSummary), sep="_")

blacklistedRegions <- import.bed(paste(analysisPath,"mm10-blacklist.v2.bed.gz",sep="/"))  
PeaksBlacklisted <- countOverlaps(subject = blacklistedRegions,
                                  query = peakSummary) != 0

table(PeaksBlacklisted)

peakSummary <- peakSummary[!PeaksBlacklisted]
```

```{r}
# define gaps between peaks
gapSummary <- gaps(peakSummary)
GapsBlacklisted <- countOverlaps(subject = blacklistedRegions,
                                 query = gapSummary) != 0
table(GapsBlacklisted)
gapSummary <- gapSummary[!GapsBlacklisted]
```

```{r}
### Dataframe of peakPresence
peakSummary.df <- as.data.frame(peakSummary@elementMetadata)
rownames(peakSummary.df)<- peakSummary@ranges@NAMES
colnames(peakSummary.df)<- unlist(lapply(strsplit(colnames(peakSummary.df),split ="_peaks"), function(x) x[[1]]))
```

###### Histogram on peak width

```{r, warning=FALSE}
ggplot(data.frame(peakSummary@ranges), aes(width)) + 
  geom_histogram(aes(y=..count..),binwidth = 10,fill="#A9DDD9")+
  scale_x_continuous(limits=c(0,200))+
  geom_text(aes(900, 200, label = paste(sum(data.frame(peakSummary@ranges)$width>1000),">",sep=""), 
                vjust = -0.5), col = "black")+
  theme_classic()
```

### Import Counts

```{r,warning=FALSE}
# count reads from bam files in peak regions
countsPerPeak <- peakCountTable(peaks = peakSummary, 
                                   path = bamFilePath, 
                                   pattern = ".final.bam$",
                                   )
#write peak counts into data frame
countsPerPeak.df <- as.data.frame(assay(countsPerPeak))
```

```{r,warning=FALSE}
# count reads from bam files in gap regions
countsPerGap <- peakCountTable(peaks = gapSummary, 
                               path = bamFilePath, 
                               pattern = ".final.bam$",
                               )

#write gap counts into data frame
countsPerGap.df <- as.data.frame(assay(countsPerGap))
```

### Export unfiltered counts

```{r}
# write.csv(countsPerPeak.df, "/home/caterina/data/analysis/counts_GEO_revision.csv")
```

## QC 

```{r}
# Count Statistics
totalPeakLength <- sum(as.numeric(width(peakSummary)))
totalPeakNumber <- length(peakSummary)
meanPeakLength <- round(totalPeakLength/totalPeakNumber,3)
totalPeakReads <- colSums(countsPerPeak.df)
meanReadsperPeak <- round(totalPeakReads/totalPeakNumber,3)
PeakReadsperBase <- totalPeakReads/totalPeakLength

# Gap statistics
totalGapLength <- sum(as.numeric(width(gapSummary)))
totalGapNumber <- length(gapSummary)
meanGapLength <- round(totalGapLength/totalGapNumber,3)
totalGapReads <- colSums(assay(countsPerGap))
meanReadsperGap <- round(totalGapReads/totalGapNumber,3)
GapReadsperBase <- totalGapReads/totalGapLength

summary.stats <- data.frame("totalPeakLength"=totalPeakLength,
                           "totalGapLength"=totalGapLength,
                            "totalPeakNumber"=totalPeakNumber,
                            "totalReads" = totalPeakReads+totalGapReads,
                            "totalPeakReads" = totalPeakReads,
                            "totalGapReads" = totalGapReads,
                            "meanReadsperPeak"= meanReadsperPeak,
                            "meanReadsperGap" = meanReadsperGap,
                            "PeakReadsperBase"=round(PeakReadsperBase,3),
                            "GapReadsperBase"= round(GapReadsperBase,3),
                            #"CoverageRatio" = round(PeakReadsperBase/GapReadsperBase,3))
                            "FRiP_score" = round(totalPeakReads/(totalPeakReads+totalGapReads),3))
idx <- match(rownames(summary.stats), sample_table$sample)
summary.stats$time <- sample_table$Time[idx]
summary.stats$group <- sample_table$Group[idx]
summary.stats$celltype <- sample_table$Celltype[idx]

summary.stats
```

### QC plots

```{r,fig.width=24,fig.height=6}
tmp <- melt(countsPerPeak.df)
idx <- match(tmp$variable,sample_table$ID)
tmp$ID <- sample_table$ID[idx]

p1 <- ggplot(tmp, aes(x= variable, y= value+1, fill= ID))+
  geom_boxplot() +
  scale_y_log10()+
  ylab("log10(Count+1)")+
  xlab("") +
  theme_classic()+
  theme(axis.text.x = element_text(size=8, angle = 45, hjust = 1),
        plot.title = element_text(size = 8, face = "bold"))

p2 <- ggplot(summary.stats, aes(x= rownames(summary.stats), y= totalPeakReads,
                                fill=rownames(summary.stats)))+
  geom_boxplot() +
  geom_jitter(aes(shape=rownames(summary.stats))) +
  ylab("Total Peak Counts")+
  xlab("") +
  theme_classic()+
  theme(axis.text.x = element_text(size=8, angle = 45, hjust = 1),
        plot.title = element_text(size = 8, face = "bold"))

# visualize the FRiP score
p3 <- ggplot(summary.stats, aes(x = rownames(summary.stats), y = FRiP_score, fill=rownames(summary.stats))) +
  geom_boxplot() +
  geom_jitter(aes(shape=rownames(summary.stats))) +
  ylab("FRiP score")+
  xlab("") +
  theme_classic()+
  theme(axis.text.x = element_text(size=8, angle = 45, hjust = 1),
        plot.title = element_text(size = 8, face = "bold"))

plot_grid(p1, p2, p3, ncol=2, rel_widths = c(2.3,1))
```
### QC board

```{r, fig.width=14, fig.height=4.5}
# format TSS-ES
TSS_persample <- t(as.data.frame(TSS_df))
colnames(TSS_persample)[1] <- "TSS_ES"
TSS_persample <- as.data.frame(TSS_persample)
TSS_persample$ID <- rownames(TSS_persample)
reduced_st <- sample_table[, c(2, 9)]
TSS_persample <- merge(reduced_st, TSS_persample, by="ID")

# format FRiP score and log10(unique_reads)
FRiP_UR <- summary.stats[, c(4,11)]
FRiP_UR$log10UR <- log10(FRiP_UR$totalReads)
FRiP_UR$ID <- rownames(FRiP_UR)
FRiP_UR <- merge(reduced_st, FRiP_UR, by="ID")

# format mito%
mito_perc <- read.csv("/home/caterina/data/analysis/lorenzo/mito_count/mito_percent.csv", header = FALSE)
mito_perc <- mito_perc[c(1:13, 15:19), ]
colnames(mito_perc)[2] <- "mito%"
mito_perc$ID <- paste0("Sample_", mito_perc$V1)
mito_perc$V1 <- NULL
mito_perc <- merge(reduced_st, mito_perc, by="ID")

# plot board as one-col heatmap

# TSS_ES
my_breaks_TSS <- c(min(TSS_persample$TSS_ES), max(TSS_persample$TSS_ES))

TSS_hm <- ggplot(TSS_persample, aes(1, SAMPLE_NAME, fill = TSS_ES)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("#e5f9ec", "#008e2f"), breaks = my_breaks_TSS) +
  geom_text(aes(label = round(TSS_ES, 2))) +
  xlab("TSS_ES") +
  ylab("") +
  theme_bw(base_size = 13) +
  theme(plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.text=element_text(size=8),
    aspect.ratio = 5)

# FRiP
my_breaks_FRiP <- c(min(FRiP_UR$FRiP_score), max(FRiP_UR$FRiP_score))

FRiP_hm <- ggplot(FRiP_UR, aes(1, SAMPLE_NAME, fill = FRiP_score)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("#f6cef6", "#7e067e"), breaks = my_breaks_FRiP) +
  geom_text(aes(label = round(FRiP_score, 2))) +
  xlab("FRiP_score") +
  ylab("") +
  theme_bw(base_size = 13) +
  theme(plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.text=element_text(size=8),
    aspect.ratio = 5)

# log10UF
my_breaks_log10UR <- c(min(FRiP_UR$log10UR), max(FRiP_UR$log10UR))

log10UR_hm <- ggplot(FRiP_UR, aes(1, SAMPLE_NAME, fill = log10UR)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("#feefd8", "#e19e39"), breaks = my_breaks_log10UR) +
  geom_text(aes(label = round(log10UR, 2))) +
  xlab("log10UR") +
  ylab("") +
  theme_bw(base_size = 13) +
  theme(plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.text=element_text(size=8),
    aspect.ratio = 5)

# mito%
my_breaks_mito<- c(min(mito_perc$`mito%`), max(mito_perc$`mito%`))

mito_perc_hm <- ggplot(mito_perc, aes(1, SAMPLE_NAME, fill = `mito%`)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("#DE3E1C", "#f8d8d1"), breaks = my_breaks_mito) +
  geom_text(aes(label = round(`mito%`, 2))) +
  xlab("mito%") +
  ylab("") +
  theme_bw(base_size = 13) +
  theme(plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.text=element_text(size=8),
    aspect.ratio = 5)

grid.arrange(TSS_hm, FRiP_hm, log10UR_hm, mito_perc_hm, ncol=4)
```

### Histogram of rowSums

```{r,fig.height=4,fig.width=8, warning=FALSE}
countStats <- data.frame(row.names=rownames(countsPerPeak.df),
                         mean=rowMeans(countsPerPeak.df),
                         sum=rowSums(countsPerPeak.df))

p4 <-ggplot(countStats, aes(mean)) + 
  geom_histogram(binwidth = 1, fill= "#A9DDD9")+
  scale_x_continuous(limits=c(0,50))+
  geom_text(aes(450, 1000, label = paste(sum(countStats$mean>500),">",sep=""), vjust = -0.5), col = "black")+
  theme_classic()

p5 <-ggplot(countStats, aes(sum)) + 
  geom_histogram(binwidth = 10, fill= "#A9DDD9")+
  scale_x_continuous(limits=c(0,200))+
  geom_text(aes(4950, 200, label = paste(sum(countStats$sum>5000),">",sep=""), vjust = -0.5), col = "black")+
  geom_vline(xintercept=10)+
  theme_classic()

plot_grid(p4, p5, labels=list("Mean","Sum"), 
          label_x = 0.5, label_y = 0.9, hjust = -0.5, vjust = -0.5)
```

### Filter Peaks

```{r}
peaks2keep <- rownames(countsPerPeak.df[rowMax(as.matrix(countsPerPeak.df))>10,])
length(peaks2keep)

# filter peaks
peakSummary.filt <- peakSummary[peaks2keep]
peakSummary.filt.df <- peakSummary.df[rownames(peakSummary.df) %in% peaks2keep,]

# filter counts
countsPerPeak.filt.df <- countsPerPeak.df[peaks2keep,]
```
### Peak Annotation

### Relative localization to nearest gene

```{r,warning=FALSE}
# complete peak set
peakAnnotation <- annotatePeak(peakSummary, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, level="gene")
peakAnnotation.df <- as.data.frame(as.GRanges(peakAnnotation))

# filtered peak set
peakAnnotation.filt <- annotatePeak(peakSummary.filt, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, level="gene")
peakAnnotation.filt.df <- as.data.frame(as.GRanges(peakAnnotation.filt))
```

##### PieChart

```{r}
plotAnnoPie(peakAnnotation.filt)
```

##### Histogram

```{r, fig.width=5, fig.height=4, warning=FALSE}
ggplot(data=peakAnnotation.filt.df, aes(x = peakAnnotation.filt.df$distanceToTSS)) + 
  geom_histogram(aes(y=..count..), colour="black", fill="white", binwidth = 50) +
  scale_x_continuous(limits=c(-2000,2000))+
  scale_y_continuous(limits=c(0,2000), oob = scales::squish)+
  labs(x="Distance to TSS (bp)", y="Peaks number")+
  theme_classic()+
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10))
```

# DESeq2 analysis of peak counts

#### DESeq2 object

```{r}
# produce DESeq data set
dds <- DESeqDataSetFromMatrix(countData = countsPerPeak.filt.df,
                              colData = sample_table,
                              design = ~ CELL_TYPE) # change the design
```

#### Calculate DESeq2 Statistics

```{r}
dds <- DESeq(dds,quiet = FALSE)
```

#### Variance Stabilization

```{r}
dds_rlog <- rlog(dds, blind = FALSE)
```

#### Principal Component Analysis

```{r, fig.height=3, fig.width=4}
plotPCA(pca_input = dds_rlog,
              ntop="all", 
              xPC=1, 
              yPC=2,
              point_size=4,
              title="",
              color="GROUP",
              anno_color=col_GROUP) +
  theme(aspect.ratio = 1)+
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10))
```

## Annotated rlog normalized count table

```{r}
#write table of vst normalized counts
rlog_matrix <- as.matrix(assay(dds_rlog))
norm <- subset(rlog_matrix, rownames(rlog_matrix) %in% rownames(peakAnnotation.filt.df))

identical(rownames(norm), rownames(peakAnnotation.filt.df))

norm_anno <- cbind(norm, peakAnnotation.filt.df)
norm_anno$peakid <- rownames(norm_anno)
```

```{r}
# add symbol annotation

mart <- useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
getBM <- getBM(attributes = c("ensembl_gene_id", "mgi_symbol", "entrezgene_id"), bmHeader = T, mart = mart)
colnames(getBM)[3] <- "geneId"
colnames(getBM)[2] <- "symbol"
colnames(getBM)[1] <- "ensId"

#check out here

getBM_overlap <- subset(getBM, getBM$geneId %in% norm_anno$geneId)
getBM_overlap <- getBM_overlap[,c(2,3)]
getBM_overlap <- unique(getBM_overlap)
getBM_overlap <- getBM_overlap[!duplicated(getBM_overlap$geneId),]
norm_anno_symbol <- subset(norm_anno, norm_anno$geneId %in% getBM_overlap$geneId)
norm_anno_symbol <- merge(norm_anno_symbol, getBM_overlap, by = "geneId")
norm_anno <- norm_anno_symbol
```

# GSVA analysis

## Canonical B cell markers

```{r}
# format gsva input
norm_anno_gsva <- norm_anno[, c(2:19, 33)]

# proxy for gene accessibility is the sum of all counts per symbol, independently from peak annotation
norm_anno_gsva <- aggregate(.~symbol, data=norm_anno_gsva, FUN=sum)

norm_anno_gsva <- norm_anno_gsva[2:12389,]
rownames(norm_anno_gsva) <- norm_anno_gsva$symbol
norm_anno_gsva$symbol <- NULL
norm_anno_gsva <- as.matrix(norm_anno_gsva)
```

```{r, fig.height=3, fig.width=3}
set.seed(1)
B_cell_markers <- list(c("Cd19", "Ms4a1", "Cd74", "Cd79a", "Cd79b", "Pax5", "Bank1", "Cd22"))

# patch
norm_anno_gsva_patch <- norm_anno_gsva[,1:10]

gsva_patch_celltype <- gsva(norm_anno_gsva_patch, B_cell_markers)
gsva_spleen_bcell_patch <- as.data.frame(t(gsva_patch_celltype))
gsva_spleen_bcell_patch$ID <- rownames(gsva_spleen_bcell_patch)
gsva_spleen_bcell_patch <- merge(gsva_spleen_bcell_patch, sample_table, by = "ID")

ggplot(gsva_spleen_bcell_patch, aes(x=GROUP, y=V1, fill=GROUP)) +
  geom_boxplot(alpha=0.8) +
  scale_fill_manual(values=c("#ce7e00", "#0e2f44")) +
  geom_jitter(position=position_jitter(0.2)) +
  stat_compare_means(method = "wilcox.test", ref.group = "T_patch", label = "p.format") +
  xlab("") +
  ylab("") +
  theme_bw() +
  theme(axis.text.x = element_text(color="black", 
                           size=10),
          axis.text.y = element_text(color="black", 
                           size=10))

# 100 cells 
norm_anno_gsva_100 <- norm_anno_gsva[,11:18]

gsva_100_celltype <- gsva(norm_anno_gsva_100, B_cell_markers)

gsva_spleen_bcell <- as.data.frame(t(gsva_100_celltype))
gsva_spleen_bcell$ID <- rownames(gsva_spleen_bcell)
gsva_spleen_bcell <- merge(gsva_spleen_bcell, sample_table, by = "ID")

ggplot(gsva_spleen_bcell, aes(x=GROUP, y=V1, fill=GROUP)) +
  geom_boxplot(alpha=0.8) +
  scale_fill_manual(values=c("#ffd966", "#468499")) +
  geom_jitter(position=position_jitter(0.2)) +
  stat_compare_means(method = "wilcox.test", ref.group = "T_100", label = "p.format") +
  xlab("") +
  ylab("") +
  theme_bw() +
  theme(axis.text.x = element_text(color="black", 
                           size=10),
          axis.text.y = element_text(color="black", 
                           size=10))
```

## Canonical T cell markers

```{r, fig.height=3, fig.width=3}
set.seed(1)
T_cell_markers <- list(c("Cd3d", "Cd3e", "Cd3g", "Tcf7", "Lef1", "Cxcr6", "Satb1"))

# patch
norm_anno_gsva_patch <- norm_anno_gsva[,1:10]

gsva_patch_celltype <- gsva(norm_anno_gsva_patch, T_cell_markers)
gsva_spleen_tcell_patch <- as.data.frame(t(gsva_patch_celltype))
gsva_spleen_tcell_patch$ID <- rownames(gsva_spleen_tcell_patch)
gsva_spleen_tcell_patch <- merge(gsva_spleen_tcell_patch, sample_table, by = "ID")

ggplot(gsva_spleen_tcell_patch, aes(x=GROUP, y=V1, fill=GROUP)) +
  geom_boxplot(alpha=0.8) +
  scale_fill_manual(values=c("#ce7e00", "#0e2f44")) +
  geom_jitter(position=position_jitter(0.2)) +
  stat_compare_means(method = "wilcox.test", ref.group = "T_patch", label = "p.format") +
  xlab("") +
  ylab("") +
  theme_bw() +
  theme(axis.text.x = element_text(color="black", 
                           size=10),
          axis.text.y = element_text(color="black", 
                           size=10))

# 100 cells 
norm_anno_gsva_100 <- norm_anno_gsva[,11:18]

gsva_100_celltype <- gsva(norm_anno_gsva_100, T_cell_markers)

gsva_spleen_tcell <- as.data.frame(t(gsva_100_celltype))
gsva_spleen_tcell$ID <- rownames(gsva_spleen_tcell)
gsva_spleen_tcell <- merge(gsva_spleen_tcell, sample_table, by = "ID")

ggplot(gsva_spleen_tcell, aes(x=GROUP, y=V1, fill=GROUP)) +
  geom_boxplot(alpha=0.8) +
  scale_fill_manual(values=c("#ffd966", "#468499")) +
  geom_jitter(position=position_jitter(0.2)) +
  stat_compare_means(method = "wilcox.test", ref.group = "T_100", label = "p.format") +
  xlab("") +
  ylab("") +
  theme_bw() +
  theme(axis.text.x = element_text(color="black", 
                           size=10),
          axis.text.y = element_text(color="black", 
                           size=10))
```

## HM of per gene accessibility (signature genes)

```{r, fig.height=4, fig.width=3}
# as in GSVA per gene accessibility defined as sum of counts from peaks mapping on that nearestGene

signatures_genes <- factor(c(T_cell_markers[[1]], B_cell_markers[[1]]))

hm_normanno <- subset(norm_anno_gsva, rownames(norm_anno_gsva) %in% signatures_genes)
hm_normanno_100 <- hm_normanno[, 11:18]
hm_normanno_patch <- hm_normanno[, 1:10]

hm_normanno_100 <- scale(t(hm_normanno_100))
hm_normanno_100 <- t(hm_normanno_100)
hm_normanno_100 <- as.data.frame(hm_normanno_100)

hm_normanno_patch <- scale(t(hm_normanno_patch))
hm_normanno_patch <- t(hm_normanno_patch)
hm_normanno_patch <- as.data.frame(hm_normanno_patch)

set.seed(1)

viridis(100)

col_fun <- colorRamp2(seq(-1.9, 1.9, length.out = 100), inferno(100))
HM_clust_100 <- Heatmap(as.matrix(hm_normanno_100),
                col = col_fun, na_col = "black",border = T,
                show_row_names = TRUE,
                cluster_columns = TRUE,
                cluster_rows = FALSE,
                use_raster = F,
                row_order = signatures_genes)
HM_clust_patch <- Heatmap(as.matrix(hm_normanno_patch),
                col = col_fun, na_col = "black",border = T,
                show_row_names = TRUE,
                cluster_columns = TRUE,
                cluster_rows = FALSE,
                use_raster = F,
                row_order = signatures_genes)

HM_clust_100
HM_clust_patch
```


# Genome browser tracks

## Cd3e

```{r, fig.height=3, fig.width=3}
afrom <- 44960136
ato <- 45055187

gtrack <- GenomeAxisTrack()

options(ucscChromosomeNames=FALSE)
rtrack <- UcscTrack(genome = "mm10", chromosome = "chr9",
                      track = "All GENCODE VM16", from = afrom, to = ato,
                      trackType = "GeneRegionTrack", 
                      rstarts = "exonStarts", rends = "exonEnds", 
                      gene = "name",  symbol = "name2", 
                      transcript = "name", strand = "strand",
                      fill = "#8282d2", 
                      name = "Other RefSeq", showID = T, geneSymbol = T)

#RPKM norm tracks

bw_B <- "/home/caterina/data/analysis/lorenzo/bw_merged/all_B/merged_sorted.bw"
bw_T <- "/home/caterina/data/analysis/lorenzo/bw_merged/all_T/merged_sorted.bw"


bwtrack_B <- DataTrack(bw_B, isPaired = T, name = "B_cells", type  = 'hist', fill = "#ce7e00", col.histogram = "NA")
bwtrack_T <- DataTrack(bw_T, isPaired = T, name = "T_cells", type  = 'hist', fill = "#0e2f44", col.histogram = "NA")

options(ucscChromosomeNames=FALSE)

plotTracks(c(bwtrack_B, bwtrack_T, gtrack, rtrack), from = afrom, to = ato, type = "hist", window = -1, windowSize = 1500, chromosome = "chr9", transcriptAnnotation = "symbol", sizes=c(1,1,1,1), legend = FALSE, col.title = "black", col.axis = "black", cex.axis = 0.5, ylim=c(0,20)) 
```

## Tcf7

```{r, fig.height=3, fig.width=3}
afrom <- 52217497
ato <- 52305070

gtrack <- GenomeAxisTrack()

options(ucscChromosomeNames=FALSE)
rtrack <- UcscTrack(genome = "mm10", chromosome = "chr11",
                      track = "All GENCODE VM16", from = afrom, to = ato,
                      trackType = "GeneRegionTrack", 
                      rstarts = "exonStarts", rends = "exonEnds", 
                      gene = "name",  symbol = "name2", 
                      transcript = "name", strand = "strand",
                      fill = "#8282d2", 
                      name = "Other RefSeq", showID = T, geneSymbol = T)

#RPKM norm tracks

bw_B <- "/home/caterina/data/analysis/lorenzo/bw_merged/all_B/merged_sorted.bw"
bw_T <- "/home/caterina/data/analysis/lorenzo/bw_merged/all_T/merged_sorted.bw"


bwtrack_B <- DataTrack(bw_B, isPaired = T, name = "B_cells", type  = 'hist', fill = "#ce7e00", col.histogram = "NA")
bwtrack_T <- DataTrack(bw_T, isPaired = T, name = "T_cells", type  = 'hist', fill = "#0e2f44", col.histogram = "NA")

options(ucscChromosomeNames=FALSE)

plotTracks(c(bwtrack_B, bwtrack_T, gtrack, rtrack), from = afrom, to = ato, type = "hist", window = -1, windowSize = 1500, chromosome = "chr11", transcriptAnnotation = "symbol", sizes=c(1,1,1,1), legend = FALSE, col.title = "black", col.axis = "black", cex.axis = 0.5, ylim=c(0,20)) 
```

## Ms4a1

```{r, fig.height=3, fig.width=3}
afrom <- 11230000
ato <- 11295915

gtrack <- GenomeAxisTrack()

options(ucscChromosomeNames=FALSE)
rtrack <- UcscTrack(genome = "mm10", chromosome = "chr19",
                      track = "All GENCODE VM16", from = afrom, to = ato,
                      trackType = "GeneRegionTrack", 
                      rstarts = "exonStarts", rends = "exonEnds", 
                      gene = "name",  symbol = "name2", 
                      transcript = "name", strand = "strand",
                      fill = "#8282d2", 
                      name = "Other RefSeq", showID = T, geneSymbol = T)

#RPKM norm tracks

bw_B <- "/home/caterina/data/analysis/lorenzo/bw_merged/all_B/merged_sorted.bw"
bw_T <- "/home/caterina/data/analysis/lorenzo/bw_merged/all_T/merged_sorted.bw"


bwtrack_B <- DataTrack(bw_B, isPaired = T, name = "B_cells", type  = 'hist', fill = "#ce7e00", col.histogram = "NA")
bwtrack_T <- DataTrack(bw_T, isPaired = T, name = "T_cells", type  = 'hist', fill = "#0e2f44", col.histogram = "NA")

options(ucscChromosomeNames=FALSE)

plotTracks(c(bwtrack_B, bwtrack_T, gtrack, rtrack), from = afrom, to = ato, type = "hist", window = -1, windowSize = 1500, chromosome = "chr19", transcriptAnnotation = "symbol", sizes=c(1,1,1,1), legend = FALSE, col.title = "black", col.axis = "black", cex.axis = 0.5, ylim=c(0,20)) 
```

## Cd19

```{r, fig.height=3, fig.width=3}
afrom <- 126387219
ato <- 126446120

gtrack <- GenomeAxisTrack()

options(ucscChromosomeNames=FALSE)
rtrack <- UcscTrack(genome = "mm10", chromosome = "chr7",
                      track = "All GENCODE VM16", from = afrom, to = ato,
                      trackType = "GeneRegionTrack", 
                      rstarts = "exonStarts", rends = "exonEnds", 
                      gene = "name",  symbol = "name2", 
                      transcript = "name", strand = "strand",
                      fill = "#8282d2", 
                      name = "Other RefSeq", showID = T, geneSymbol = T)

#RPKM norm tracks

bw_B <- "/home/caterina/data/analysis/lorenzo/bw_merged/all_B/merged_sorted.bw"
bw_T <- "/home/caterina/data/analysis/lorenzo/bw_merged/all_T/merged_sorted.bw"


bwtrack_B <- DataTrack(bw_B, isPaired = T, name = "B_cells", type  = 'hist', fill = "#ce7e00", col.histogram = "NA")
bwtrack_T <- DataTrack(bw_T, isPaired = T, name = "T_cells", type  = 'hist', fill = "#0e2f44", col.histogram = "NA")

options(ucscChromosomeNames=FALSE)

plotTracks(c(bwtrack_B, bwtrack_T, gtrack, rtrack), from = afrom, to = ato, type = "hist", window = -1, windowSize = 1500, chromosome = "chr7", transcriptAnnotation = "symbol", sizes=c(1,1,1,1), legend = FALSE, col.title = "black", col.axis = "black", cex.axis = 0.5, ylim=c(0,20))  
```

# ChromVAR 

## Get GC content

```{r}
set.seed(1)
head(countsPerPeak)

# get GC content of peaks to determine background peaks with use of addGCBias function
# output: updated summarized experiment with new column(bias)
mouse <- BSgenome.Mmusculus.UCSC.mm10

# chromosomes do not match
seqlevels(countsPerPeak)
seqlevels(mouse)

# filter counts
counts_filtered <- countsPerPeak[peaks2keep]

# subset genome and counts for standard chromosomes
chromo <- c("chr1","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chrM","chrX","chrY")
mouse <- getSeq(mouse, chromo)
new_counts <- keepSeqlevels(counts_filtered, chromo)

#find indices of samples to keep

counts_bias <- addGCBias(new_counts, 
                          genome = mouse)

head(rowData(counts_bias))
```

## Get motifs 

```{r}
set.seed(1)

# motifs consensus from ATACseqTFEA
# https://bioconductor.org/packages/devel/bioc/vignettes/ATACseqTFEA/inst/doc/ATACseqTFEA.html
# motifs <- readRDS(system.file("extdata", "PWMatrixList.rds",
#                                package="ATACseqTFEA"))

motifs_ATACseq_TFEA <- readRDS("/home/caterina/data/analysis/hannah/motifs_ATACseq_TFEA.rds")

# find motifs within peaks via motifmatchr 

motif_ix <- matchMotifs(motifs_ATACseq_TFEA, counts_bias,
                        genome = mouse)
```

## Compute deviations 

```{r}
set.seed(1)
#get deviation -> representation of how accessible peaks are relative to expectation based on equal chromatin accessibility profiles normalized by background peaks 

dev <- computeDeviations(object = counts_bias, annotations = motif_ix)
```

corri da qui sotto

## Variability 

```{r}
set.seed(1)
# access variability 

variability <- computeVariability(dev)
plotVariability(variability, use_plotly = FALSE)
```

## Convert dev into df

```{r}
# dev contains the bias corrected deviations (dev) and the deviation z score (z) of all samples for all TF motifs 

x <- assays(dev)$z
y <- assays(dev)$deviations
dev.df <- as.data.frame(y, row.names = assays(dev)$NAMES, 
                        cut.names = FALSE, col.names =names (y), fix.empty.names = TRUE,
                        stringsAsFactors = FALSE)

z.df <- as.data.frame(x, row.names = assays(dev)$NAMES, 
                        cut.names = FALSE, col.names =names (x), fix.empty.names = TRUE,
                        stringsAsFactors = FALSE)
```

## Calculate median z-score

```{r}
# Calculate median for all T and B cell samples

z.df$B_median = rowMedians(as.matrix(z.df[,c(1, 3, 6, 8, 10, 11, 12, 13, 14)])) 

z.df$T_median = rowMedians(as.matrix(z.df[,c(2, 4, 5, 7, 9, 15, 16, 17, 18)])) 
```

## B vs T delta

```{r}
z.df$delta_median <- z.df$T_median - z.df$B_median
```

## Boxplot

```{r, fig.height=3, fig.width=3}
anno_z.df <- as.data.frame(t(z.df))
anno_z.df$ID <- rownames(anno_z.df)
anno_z.df <- merge(anno_z.df, sample_table[, c(3,8,9)], by="ID")
rownames(anno_z.df) <- anno_z.df$ID
anno_z.df$ID <- NULL

# T cells: GATA3

tfoi_gata3 <- anno_z.df[, c("GATA3", "CELL_TYPE", "GROUP")]

ggboxplot(tfoi_gata3, x = "CELL_TYPE", y = "GATA3",
          ylab = "z-score") +
          geom_jitter(position=position_jitter(0.2), pch=21, aes(fill=factor(GROUP)), size=2.7) +
          stat_compare_means(method = "wilcox.test", label = "p.format")+
          scale_fill_manual(values=c("#ffd966", "#ce7e00", "#468499", "#0e2f44"))+
          theme_bw() +
          theme(axis.text.x = element_text(color="black",
                           size=10),
          axis.text.y = element_text(color="black",
                           size=10))

# B cells: EBF1

tfoi_ebf1 <- anno_z.df[, c("EBF1", "CELL_TYPE", "GROUP")]

ggboxplot(tfoi_ebf1, x = "CELL_TYPE", y = "EBF1",
          ylab = "z-score") +
          geom_jitter(position=position_jitter(0.2), pch=21, aes(fill=factor(GROUP)), size=2.7) +
          stat_compare_means(method = "wilcox.test", label = "p.format")+
          scale_fill_manual(values=c("#ffd966", "#ce7e00", "#468499", "#0e2f44"))+
          theme_bw() +
          theme(axis.text.x = element_text(color="black",
                           size=10),
          axis.text.y = element_text(color="black",
                           size=10))
```

## Top 25 enriched TFs for B and T cells (highest z-scores)

```{r}
# order the medians of all T and B cells

B_ordered <- z.df[order(z.df$B_median, decreasing = TRUE),]
T_ordered <- z.df[order(z.df$T_median, decreasing = TRUE),]

B_top <- rownames(B_ordered[1:25, ])
T_top <- rownames(T_ordered[1:25, ])

top <- c(B_top, T_top)

top_df <- subset(z.df, rownames(z.df) %in% top)
top_df <- top_df[,1:18]

top_df <- scale(t(top_df))
top_df <- t(top_df)
```

## Plot HM of top 20 enriched TFs

```{r, fig.height=6.5, fig.width=4.5}
set.seed(1)
ann <- data.frame(sample_table$GROUP)
colnames(ann) <- c('GROUP')

colAnn <- HeatmapAnnotation(df = ann,
  col = anno_color,
  annotation_width = unit(c(1, 4), 'cm'),
  gap = unit(1, 'mm'))

SamplesPatch_heatmap <- Heatmap(top_df, cluster_rows = TRUE, cluster_columns = TRUE,
         name = "z-score", #title of legend
        row_title = "TF motifs",
         row_names_gp = gpar(fontsize = 7), # Text size for row names
         top_annotation=colAnn
         )

SamplesPatch_heatmap
```

## Gata3

```{r, fig.height=3, fig.width=3}
afrom <- 9813358
ato <- 9922320

gtrack <- GenomeAxisTrack()

options(ucscChromosomeNames=FALSE)
rtrack <- UcscTrack(genome = "mm10", chromosome = "chr2",
                      track = "All GENCODE VM16", from = afrom, to = ato,
                      trackType = "GeneRegionTrack", 
                      rstarts = "exonStarts", rends = "exonEnds", 
                      gene = "name",  symbol = "name2", 
                      transcript = "name", strand = "strand",
                      fill = "#8282d2", 
                      name = "Other RefSeq", showID = T, geneSymbol = T)

#RPKM norm tracks

bw_B <- "/home/caterina/data/analysis/lorenzo/bw_merged/all_B/merged_sorted.bw"
bw_T <- "/home/caterina/data/analysis/lorenzo/bw_merged/all_T/merged_sorted.bw"


bwtrack_B <- DataTrack(bw_B, isPaired = T, name = "B_cells", type  = 'hist', fill = "#ce7e00", col.histogram = "NA")
bwtrack_T <- DataTrack(bw_T, isPaired = T, name = "T_cells", type  = 'hist', fill = "#0e2f44", col.histogram = "NA")

options(ucscChromosomeNames=FALSE)

plotTracks(c(bwtrack_B, bwtrack_T, gtrack, rtrack), from = afrom, to = ato, type = "hist", window = -1, windowSize = 1500, chromosome = "chr2", transcriptAnnotation = "symbol", sizes=c(1,1,1,1), legend = FALSE, col.title = "black", col.axis = "black", cex.axis = 0.5, ylim=c(0,20)) 
```

## Ebf1

```{r, fig.height=3, fig.width=3}
afrom <- 44373082
ato <- 45252325

gtrack <- GenomeAxisTrack()

options(ucscChromosomeNames=FALSE)
rtrack <- UcscTrack(genome = "mm10", chromosome = "chr11",
                      track = "All GENCODE VM16", from = afrom, to = ato,
                      trackType = "GeneRegionTrack", 
                      rstarts = "exonStarts", rends = "exonEnds", 
                      gene = "name",  symbol = "name2", 
                      transcript = "name", strand = "strand",
                      fill = "#8282d2", 
                      name = "Other RefSeq", showID = T, geneSymbol = T)

#RPKM norm tracks

bw_B <- "/home/caterina/data/analysis/lorenzo/bw_merged/all_B/merged_sorted.bw"
bw_T <- "/home/caterina/data/analysis/lorenzo/bw_merged/all_T/merged_sorted.bw"


bwtrack_B <- DataTrack(bw_B, isPaired = T, name = "B_cells", type  = 'hist', fill = "#ce7e00", col.histogram = "NA")
bwtrack_T <- DataTrack(bw_T, isPaired = T, name = "T_cells", type  = 'hist', fill = "#0e2f44", col.histogram = "NA")

options(ucscChromosomeNames=FALSE)

plotTracks(c(bwtrack_B, bwtrack_T, gtrack, rtrack), from = afrom, to = ato, type = "hist", window = -1, windowSize = 1500, chromosome = "chr11", transcriptAnnotation = "symbol", sizes=c(1,1,1,1), legend = FALSE, col.title = "black", col.axis = "black", cex.axis = 0.5, ylim=c(0,20)) 
```

# Session info

```{r}
sessionInfo <- sessionInfo()
```


