---
title: "LCM_ATAC_mouse_revision"
author: "C Carraro"
date: "29/06/23"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Install packages

```{r}
list.of.bioc.packages<- c("Rsamtools",
                          "GenomicAlignments",
                          "rtracklayer",
                          "ChIPseeker",
                          "DBI",
                          "TxDb.Mmusculus.UCSC.mm10.knownGene",
                          "org.Mm.eg.db",
                          "DESeq2",
                          "org.Hs.eg.db",
                          "vsn",
                          "tximport",
                          "rhdf5",
                          "clusterProfiler",
                          "DOSE",
                          "GSEABase",
                          "RColorBrewer",
                          "ComplexHeatmap",
                          "pheatmap",
                          "genefilter",
                          "biomaRt",
                          "limma",
                          "sva",
                          "IHW",
                          "Gviz",
                          "tidyverse",
                          "useful",
                          "gplots",
                          "SummarizedExperiment",
                          "BiocParallel",
                          "ggrepel",
                          "ggbeeswarm",
                          "grid",
                          "gridextra",
                          "reshape2",
                          "factoextra",
                          "Hmisc",
                          "GenomicRanges",
                          "BSgenome.Mmusculus.UCSC.mm10",
                          "cowplot",
                          "HelloRanges",
                          "ATACseqQC",
                          "BRGenomics")
new.packages.bioc <- list.of.bioc.packages[!(list.of.bioc.packages %in% installed.packages()[,"Package"])]
 
if(length(new.packages.bioc)>0)if (!requireNamespace("BiocManager")) install.packages("BiocManager")
BiocManager::install(new.packages.bioc, update = FALSE)
```

### Load packages

```{r, message=FALSE}
library(circlize)
library(viridis)
library(viridisLite)
library(tidyverse)
library(useful)
library(gplots)
library(SummarizedExperiment)
library(BiocParallel)
library(ggrepel)
library(grid)
library(gridExtra)
library(reshape2)
library(factoextra)
library(Hmisc)
library(GenomicRanges)
library(GenomicAlignments)
library(rtracklayer)
library(Rsamtools)
library(ChIPseeker)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(biomaRt)
library(org.Mm.eg.db) 
library(org.Hs.eg.db) 
library(DESeq2)
library(tximport)
library(vsn)
library(IHW)
library(rhdf5)
library(genefilter)
library(fdrtool)
library(RColorBrewer)
library(pheatmap)
library(cowplot) 
library(Gviz)
library(GSEABase)
library(clusterProfiler)
library(DOSE)
library(sva)
library(limma)
library(BSgenome.Mmusculus.UCSC.mm10)
library(UpSetR)
library(HelloRanges)
library(ComplexHeatmap)
library(ATACseqQC)
library(BRGenomics)
```

# ATAC-seq analysis

## Functions

### Peak import

```{r}
# Importing of BED files and production of GRanges object of all peaks
# Modified to support multiple files exclusion
importBEDpeaks <- function(pattern = NULL, 
                           path = NULL,
                           exclude = NULL){
  fileNames <- dir(path, pattern = pattern)
  if(!is.null(exclude)){
    for (i in exclude){
  fileNames <- fileNames[!grepl(i, fileNames)] 
  }
  }
  bedPeaks <- GRangesList()
  for (i in fileNames) {
    tmp <- read.table(paste(path,i,sep="/"))
    bedPeaks[[i]] <- GRanges(seqnames = Rle(tmp$V1), ranges = IRanges(start = tmp$V2, end = tmp$V3), strand = Rle("*"))
    bedPeaks[[i]] <- bedPeaks[[i]][grepl("chr",bedPeaks[[i]]@seqnames),]
  }
  if (!is.null(pattern)) names(bedPeaks) <- sub(pattern, "", names(bedPeaks), fixed = TRUE)
  return(bedPeaks)
}
```

### Peak unification

```{r}
# Produce union of all peaks for counting of the reads
unionPeaks <- function(peaks) {
  upks <- GenomicRanges::reduce(unlist(peaks))
  for (i in names(peaks)) {
    mcols(upks)[, i] <- countOverlaps(subject = peaks[[i]],
                                      query = upks) != 0
  }
  names(upks) <- paste(seqnames(upks), start(upks), end(upks), sep="_")
  return(upks)
}
```

### Count table

```{r}
# Prepare count table for all peaks
peakCountTable <- function(peaks, path = NULL, pattern = NULL,
                           exclude = NULL) {
  fileNames <- dir(path, pattern = pattern)
  if(!is.null(exclude)){
    for (i in exclude){
  fileNames <- fileNames[!grepl(i, fileNames)] 
  }
  }
  bamFiles <- BamFileList(paste(path,fileNames,sep="/"), asMates = TRUE)
  countTable <- summarizeOverlaps(features = peaks,
                                  reads = bamFiles,
                                  mode = "Union",
                                  singleEnd = FALSE,
                                  fragments =FALSE,
                                  ignore.strand = TRUE)
  if (!is.null(pattern)) colnames(countTable) <- sub(pattern, "", colnames(countTable))
  return(countTable)
}
```

### PCA

```{r}
plotPCA <- function(pca_input = dds_vst,
                    ntop="all", 
                    xPC=1, 
                    yPC=2,
                    point_size=3,
                    title="",
                    color,
                    #shape,
                    anno_color="NULL"){
  
  vst_matrix <- as.matrix(assay(pca_input))
  
  if(ntop=="all"){
    pca <- prcomp(t(vst_matrix)) 
  }else{
    # select the ntop genes by variance
    select <- order(rowVars(vst_matrix), decreasing=TRUE)[c(1:ntop)]
    pca <- prcomp(t(vst_matrix[select,]))
  }
  
  #calculate explained variance per PC
  explVar <- pca$sdev^2/sum(pca$sdev^2)
  # transform variance to percent
  percentVar <- round(100 * explVar[c(xPC,yPC)], digits=1)
  
  # Define data for plotting  
  pcaData <- data.frame(xPC=pca$x[,xPC], 
                        yPC=pca$x[,yPC], 
                        color = sample_table[[color]],
                        #shape = sample_table[[shape]],
                        name= as.character(sample_table$sample),
                        stringsAsFactors = F)
  
  #plot PCA
  pca_plot <- ggplot(pcaData, aes(x = xPC, y = yPC, fill=color)) +
    geom_point(size =point_size, shape=21, color="black") + theme_bw()+
      xlab(paste0("PC ",xPC, ": ", percentVar[1], "% variance")) +
    ylab(paste0("PC ",yPC,": ", percentVar[2], "% variance"))
  
  if(anno_color[1] == "NULL"){
    pca_plot <- pca_plot + scale_color_discrete(name=color)
  }else{
    pca_plot <- pca_plot + scale_fill_manual(values=anno_color)
  }
  
  #pca_plot <- pca_plot + 
    #geom_text_repel(aes(label = name), colour = "black")+
    #xlab(paste0("PC ",xPC, ": ", percentVar[1], "% variance")) +
    #ylab(paste0("PC ",yPC,": ", percentVar[2], "% variance")) +
    #coord_fixed()+
    #theme_classic()+        
    #theme(aspect.ratio = 1)+ 
    #ggtitle(title)
  
  pca_plot
}
```

### DESeq2 output

```{r}
# Specify structure of DESeq2_analysis_object

setClass(Class = "DESeq2_analysis_object",
         slots = c(results="data.frame", DE_genes="list", Number_DE_genes="list"))


# Wrapper Function to perform DESeq2 differential testing
DEAnalysis <- function(condition,
                       alpha = 0.05, 
                       lfcThreshold = 0,
                       sigFC = 2, 
                       multiple_testing = "IHW",
                       independentFiltering="TRUE",
                       shrinkage = TRUE,
                       shrinkType = "normal"){
  # create results_list  
  results_list <- list()
  # print parameters
  results_list$parameters <-list(multiple_testing = multiple_testing,
                                 p_value_threshold = alpha,
                                 log2_FC_threshold = lfcThreshold,
                                 shrinkage = shrinkage,
                                 shrinkage_type = shrinkType)
  # Run results() function on comparisons defined in comparison table
  for (i in 1:nrow(comparison_table)){
    # create DE_object
    DE_object <- new(Class = "DESeq2_analysis_object")
    # IHW
    if (multiple_testing=="IHW") {
      res_deseq_lfc <- results(dds,
                               contrast = c(condition,
                                            paste(comparison_table$comparison[i]),
                                            paste(comparison_table$control[i])),
                               lfcThreshold = lfcThreshold,
                               alpha = alpha,
                               filterFun = ihw,
                               altHypothesis = "greaterAbs")
      # Independent Filtering
    }else {
      res_deseq_lfc <- results(dds,
                               contrast = c(condition,
                                            paste(comparison_table$comparison[i]),
                                            paste(comparison_table$control[i])),
                               lfcThreshold = lfcThreshold,
                               alpha = alpha,
                               independentFiltering = independentFiltering,
                               altHypothesis = "greaterAbs",
                               pAdjustMethod= multiple_testing)
    }
    if(shrinkage == TRUE){
      res_deseq_lfc <- lfcShrink(dds, 
                                 contrast = c(condition,
                                              paste(comparison_table$comparison[i]),
                                              paste(comparison_table$control[i])),
                                 res=res_deseq_lfc,
                                 type = shrinkType)
    }
    res_deseq_lfc <- as.data.frame(res_deseq_lfc)
    # indicate significant DE genes  
    res_deseq_lfc$regulation <- ifelse(!is.na(res_deseq_lfc$padj)&
                                         res_deseq_lfc$padj <= alpha&
                                         res_deseq_lfc$log2FoldChange > log(sigFC,2),
                                       "up",
                                       ifelse(!is.na(res_deseq_lfc$padj)&
                                                res_deseq_lfc$padj <= alpha&
                                                res_deseq_lfc$log2FoldChange < -log(sigFC,2),
                                              "down",
                                              "n.s."))
    # add gene annotation to results table
    res_deseq_lfc$peakid <- row.names(res_deseq_lfc) # ensembl-IDs as row names
    res_deseq_lfc <- merge(res_deseq_lfc, 
                           norm_anno[,c("peakid", 
                                        "symbol",
                                        "geneChr",
                                        "geneStart",
                                        "geneEnd",
                                        "geneLength",
                                        "geneStrand",
                                        "geneId",
                                        "distanceToTSS",
                                        "start",
                                        "end",
                                        "annotation")], 
                           by = "peakid") 
    row.names(res_deseq_lfc) <- res_deseq_lfc$peakid
    res_deseq_lfc$comparison<-paste(comparison_table$comparison[i]," vs ",comparison_table$control[i],
                                    sep="")
    # re-order results table
    if (multiple_testing=="IHW") {
      res_deseq_lfc<-res_deseq_lfc[,c("peakid", 
                                        "symbol",
                                      "geneId",
                                        "geneChr",
                                     "comparison",
                                     "regulation",
                                     "baseMean",
                                     "log2FoldChange",
                                     "lfcSE",
                                     "stat",
                                     "pvalue",
                                     "padj",
                                     "weight",
                                     "annotation")]
    }else{
      res_deseq_lfc<-res_deseq_lfc[,c("peakid", 
                                        "symbol",
                                      "geneId",
                                        "geneChr",
                                      "comparison",
                                      "regulation",
                                      "baseMean",
                                      "log2FoldChange",
                                      "lfcSE",
                                      "stat",
                                      "pvalue",
                                      "padj",
                                      "annotation")]
    }
    # print result table
    DE_object@results <- res_deseq_lfc
    # print DE genes in seperate tables
    DE_object@DE_genes <- list(up_regulated_Genes = res_deseq_lfc[res_deseq_lfc$regulation =="up",],
                               down_regulated_Genes= res_deseq_lfc[res_deseq_lfc$regulation =="down",])
    # print the numbers of DE genes
    DE_object@Number_DE_genes <- list(up_regulated_Genes = nrow(DE_object@DE_genes$up_regulated_Genes),
                                      down_regulated_Genes= nrow(DE_object@DE_genes$down_regulated_Genes))
    # write DE_object into results_list
    results_list[[paste(comparison_table$comparison[i], "vs", comparison_table$control[i], sep="_")]] <- DE_object
  }
  return(results_list)
}
```

### Union of DAR (all)

```{r}
uDARall <- function(comparisons){
  uDARs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DARs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation %in% c("up","down"),])
    uDARs <- unique(c(uDARs, DARs$peakid))
  }
  uDARs
}
```

### Union of DAR (up)

```{r}
uDARall_up <- function(comparisons){
  uDARs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DARs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation == "up",])
    uDARs <- unique(c(uDARs, DARs$peakid))
  }
  uDARs
}
```

### Union of DAR (down)

```{r}
uDARall_down <- function(comparisons){
  uDARs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DARs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation == "down",])
    uDARs <- unique(c(uDARs, DARs$peakid))
  }
  uDARs
}
```

### Union of DAR (promoter regions)

```{r}
uDAR_promo <- function(comparisons){
  uDARs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DARs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation %in% c("up","down"),])
    DARs <- subset(DARs, DARs$annotation %in% c("Promoter (<=1kb)", "Promoter (1-2kb)", "Promoter (2-3kb)"))
    uDARs <- unique(c(uDARs, DARs$peakid))
  }
  uDARs
}
```

### Union of DAR up (promoter regions)

```{r}
uDAR_promo_up <- function(comparisons){
  uDARs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DARs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation == "up",])
    DARs <- subset(DARs, DARs$annotation %in% c("Promoter (<=1kb)", "Promoter (1-2kb)", "Promoter (2-3kb)"))
    uDARs <- unique(c(uDARs, DARs$peakid))
  }
  uDARs
}
```

### Union of DAR down (promoter regions)

```{r}
uDAR_promo_down <- function(comparisons){
  uDARs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DARs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation == "down",])
    DARs <- subset(DARs, DARs$annotation %in% c("Promoter (<=1kb)", "Promoter (1-2kb)", "Promoter (2-3kb)"))
    uDARs <- unique(c(uDARs, DARs$peakid))
  }
  uDARs
}
```

### Union of DEgenes (promoter regions)

```{r}
uDEG <- function(comparisons){
  uDEGs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DEGs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation %in% c("up","down"),])
    DEGs <- subset(DEGs, DEGs$annotation %in% c("Promoter (<=1kb)", "Promoter (1-2kb)", "Promoter (2-3kb)"))
    uDEGs <- unique(c(uDEGs, DEGs$symbol))
  }
  uDEGs
}
```

### Union of DEgenes UP (promoter regions)

```{r}
uDEGup <- function(comparisons){
  uDEGs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DEGs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation %in% c("up"),])
  DEGs <- subset(DEGs, DEGs$annotation %in% c("Promoter (<=1kb)", "Promoter (1-2kb)", "Promoter (2-3kb)"))
    uDEGs <- unique(c(uDEGs, DEGs$symbol))
  }
  uDEGs
}
```

### Union of DEgenes DOWN (promoter regions)

```{r}
uDEGdown <- function(comparisons){
  uDEGs <- NULL
  tmp <- DEresults[names(DEresults) %in% comparisons]
  for(i in 1:length(comparisons)){
    DEGs <- as.data.frame(tmp[[i]]@results[tmp[[i]]@results$regulation %in% c("down"),])
    DEGs <- subset(DEGs, DEGs$annotation %in% c("Promoter (<=1kb)", "Promoter (1-2kb)", "Promoter (2-3kb)"))
    uDEGs <- unique(c(uDEGs, DEGs$symbol))
  }
  uDEGs
}
```

###GSEA dotplot for multiple comparisons

```{r}
dotplotGSEA_MC <- function(x,
                        show=25,
                        font.size=10,
                        title.size=10,
                        title.width=100){
    ggplot(x, aes(x = Cluster, y = Description, color = qvalue)) +
      geom_point(aes(size = Count)) +
      scale_x_discrete(drop = FALSE) +
      scale_colour_gradientn(colours=c('darkorange1',
                                       '#FFFF0000',
                                       '#FFFF0000'),
                             limits=c(0,1),
                             values   = c(0,0.25,0.26,1),
                             breaks   = c(0.25,0.26,1),
                             labels = format(c(0.25,0.26,1))) +
      ylab(NULL) +
      # ggtitle(paste(strwrap(unique(x$Enrichment), width=title.width), collapse = "\n"))+
      theme_bw() +
      theme(text = element_text(size=font.size),
            plot.title = element_text(size=title.size), axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
  }
```


## Directories

```{r}
projectPath <- "/home/caterina/data/alignment/2022-10-26_lung_brain_nobulk/"
bamFilePath <- paste(projectPath,"output/mapped", sep = "")
bedFilePath <- paste(projectPath, "output/peaks", sep = "")

analysisPath <- "/home/caterina/data/analysis/"
```

## Sample Annotation 

```{r}
sample_table <- read.csv("/home/caterina/data/analysis/sample_table_subset.csv",sep=";",header=TRUE, stringsAsFactors = TRUE)

# Only brain and lung 50 and 200 cells, exclude low quality

sample_table$precise_ID <- sample_table$ID
sample_table$ID <- paste("Sample_",sample_table$precise_ID,sep="")
rownames(sample_table) <- sample_table$ID
sample_table <- sample_table[c(18:31, 33:35),]
sample_table <- sample_table[c(1:6, 10, 11, 13, 14),]
```

### Colour scheme

For hex colors use: https://www.color-hex.com

```{r}
# Tissue

col_Tissue <- c("#990000", "#7300E5")

names(col_Tissue) <- c("Lung", "Brain") 

# Condition

col_condition = c("#ad3232", "#cc7f7f", "#2e005b", "#9d4cec")

names(col_condition) <- c("LU50SCA", "LU10BPA", "BR50SCA", "BR10BPA")

# Cell_number

col_cell_number = c("#feda75", "#fa7e1e")

names(col_cell_number) <- c("50", "200")

# combine color code into list

ann_colors <- list(Tissue = col_Tissue, 
                  condition = col_condition, 
                  cell_number = col_cell_number)
```

## Data Import, QC and Processing

### TSS enrichment score

```{r}
set.seed(1)

txs <- transcripts(TxDb.Mmusculus.UCSC.mm10.knownGene)
TSS_df <- list()

for(i in sample_table$ID){
  
  message(dir(path=bamFilePath, pattern = paste0(i, ".final.bam$")))
  sortedBAM <- dir(path=bamFilePath, paste0(i, ".final.bam$"),full.names = TRUE)
  #sample <- paste(strsplit(dir(path=bamFilePath, pattern = ".final.bam$")[i],"\\.")[[1]][1],sep="")
  
  bam <- import_bam(sortedBAM,
                  mapq = 20, #Filter reads by a minimum MAPQ score. This is the correct way to filter multi-aligners.
                  revcomp = FALSE, #Logical indicating if aligned reads should be reverse-complemented.
                  shift = 0L,
                  trim.to = c("whole", "5p", "3p", "center"), #Option for selecting specific bases from the reads, applied after the revcomp and shift options. By default, the entire read is maintained.
                  ignore.strand = FALSE, #Logical indicating if the strand information should be discarded.
                  field = "score", #Metadata field name to use for readcounts, usually "score". If set to NULL, identical reads (or identical positions if trim.to options applied) are not combined, and the length of the output GRanges will be equal to the number of input reads.
                  paired_end = TRUE)

  bam_gal <- as(bam, "GAlignments")

  tsse <- TSSEscore(bam_gal, txs)
  
  TSS_df[[paste0(i)]] <- tsse$TSSEscore
  
}
```

### Insert size

```{r,fig.height=24,fig.width=12}
plotList_2 <- list()

for(i in 1:length(sample_table$ID)){
  message(dir(path=bamFilePath, pattern = ".final.bam$")[i])
  sortedBAM <- dir(path=bamFilePath, pattern = ".final.bam$",full.names = TRUE)[i]
  sample <- paste(strsplit(dir(path=bamFilePath, pattern = ".final.bam$")[i],"\\.")[[1]][1],sep="")
  
  atacReads <- readGAlignmentPairs(sortedBAM, 
                                   param = ScanBamParam(mapqFilter = 1,
                                                        flag = scanBamFlag(isPaired = TRUE, isProperPair = TRUE), 
                                                        what = c("qname", 
                                                                 "mapq", 
                                                                 "isize")))
  
  atacReads_read1 <- GenomicAlignments::first(atacReads)
  insertSizes <- abs(elementMetadata(atacReads_read1)$isize)
  
  inserts <- table(insertSizes) %>% 
    data.frame %>% 
    mutate(insertSizes = as.numeric(as.vector(insertSizes)),
           Freq = as.numeric(as.vector(Freq)))
  
  plotList_2[[i]] <- ggplot(inserts, aes(x = insertSizes, y = Freq)) + 
    geom_line()+ 
    geom_vline(xintercept = c(180, 247), colour = "red") + 
    geom_vline(xintercept = c(315, 437), colour = "darkblue") + 
    geom_vline(xintercept = c(100), colour = "darkgreen") + 
    ggtitle(sample)+
    theme_bw()
}

plot_grid(plotlist = plotList_2,ncol=2)
```

#### Overlayed FSD

```{r, fig.width=5, fig.height=4}
inserts <- list()

for(i in 1:length(plotList_2)){
  
  m <- plotList_2[[i]][["data"]]
  q <- plotList_2[[i]][["labels"]][["title"]]
  inserts[[paste0(q)]] <- m
  
}

inserts_df <- data.frame()

for(i in names(inserts)){
  
  n <- inserts[[i]]
  n$df <- paste0(i)
  
  inserts_df <- rbind(inserts_df, n)
  
}

inserts_df$df <- factor(inserts_df$df, levels = c("Sample_15838", "Sample_15839", "Sample_15840", "Sample_15841", "Sample_15842", "Sample_15843", "Sample_15847", "Sample_15848", "Sample_15850", "Sample_15851"))

fsd_n <- ggplot(inserts_df, aes(x = insertSizes, y = Freq)) + 
  geom_line(aes(color = df)) +
  xlim(0, 800) +
  ylim(0, 300000) +
  labs(x="Fragment size (bp)", y="Fragments number") +
  theme_classic() +
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10)) +
  theme(aspect.ratio = 1)

insert_subset <- inserts_df
insert_subset <- insert_subset %>% group_by(df) %>% mutate(sum = sum(Freq))
insert_subset$ratio <- ((insert_subset$Freq)*100)/(insert_subset$sum)

fsd_ratio <- ggplot(insert_subset, aes(x = insertSizes, y = ratio)) + 
  geom_line(aes(color = df)) +
  xlim(0, 800) +
  ylim(0, 1.5) +
  labs(x="Fragment size (bp)", y="Fragments %") +
  theme_classic() +
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10)) +
  theme(aspect.ratio = 1)

fsd_n
fsd_ratio
```

#### Plot groupwise

```{r}
insert_subset_lung50 <- subset(insert_subset, insert_subset$df %in% c("Sample_15838", "Sample_15839", "Sample_15840"))

insert_subset_lungBA <- subset(insert_subset, insert_subset$df %in% c("Sample_15841", "Sample_15842", "Sample_15843"))

insert_subset_brain50 <- subset(insert_subset, insert_subset$df %in% c("Sample_15847", "Sample_15848"))

insert_subset_brainBA <- subset(insert_subset, insert_subset$df %in% c("Sample_15850", "Sample_15851"))

fsd_lung50 <- ggplot(insert_subset_lung50, aes(x = insertSizes, y = ratio)) + 
  geom_line(aes(color = df)) +
  scale_colour_manual(values=c("#4C0000", "#FFB2B2", "#FF6666")) +
  xlim(0, 800) +
  ylim(0, 1.5) +
  labs(x="Fragment size (bp)", y="Fragments %") +
  theme_classic() +
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10)) +
  theme(aspect.ratio = 1) +
  ggtitle("lung50")

fsd_lungBA <- ggplot(insert_subset_lungBA, aes(x = insertSizes, y = ratio)) + 
  geom_line(aes(color = df)) +
  scale_colour_manual(values=c("#4C0000", "#FFB2B2", "#FF6666")) +
  xlim(0, 800) +
  ylim(0, 1.5) +
  labs(x="Fragment size (bp)", y="Fragments %") +
  theme_classic() +
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10)) +
  theme(aspect.ratio = 1) +
  ggtitle("lungBA")

fsd_brain50 <- ggplot(insert_subset_brain50, aes(x = insertSizes, y = ratio)) + 
  geom_line(aes(color = df)) +
  scale_colour_manual(values=c("#800080", "#BF7FBF")) +
  xlim(0, 800) +
  ylim(0, 1.5) +
  labs(x="Fragment size (bp)", y="Fragments %") +
  theme_classic() +
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10)) +
  theme(aspect.ratio = 1) +
  ggtitle("brain50")

fsd_brainBA <- ggplot(insert_subset_brainBA, aes(x = insertSizes, y = ratio)) + 
  geom_line(aes(color = df)) +
  scale_colour_manual(values=c("#800080", "#BF7FBF")) +
  xlim(0, 800) +
  ylim(0, 1.5) +
  labs(x="Fragment size (bp)", y="Fragments %") +
  theme_classic() +
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10)) +
  theme(aspect.ratio = 1) +
  ggtitle("brainBA")

grid.arrange(fsd_lung50, fsd_lungBA, fsd_brain50, fsd_brainBA, ncol=2, nrow=2)
```

### Import merged bam-called peaks

Peaks are here called on bed derived from merging all bam files (lung and brain samples). 

```{r}
tmp <- read.table(paste(bedFilePath, "merged_peaks.narrowPeak", sep="/"))
peakSummary <- GRanges(seqnames = Rle(tmp$V1), ranges = IRanges(start = tmp$V2, end = tmp$V3), strand = Rle("*"))
peakSummary <- peakSummary[grepl("chr",peakSummary@seqnames),]
names(peakSummary) <- paste(seqnames(peakSummary), start(peakSummary), end(peakSummary), sep="_")

blacklistedRegions <- import.bed(paste(analysisPath,"mm10-blacklist.v2.bed.gz",sep="/"))  
PeaksBlacklisted <- countOverlaps(subject = blacklistedRegions,
                                  query = peakSummary) != 0

table(PeaksBlacklisted)

peakSummary <- peakSummary[!PeaksBlacklisted]
```

```{r}
# define gaps between peaks
gapSummary <- gaps(peakSummary)
GapsBlacklisted <- countOverlaps(subject = blacklistedRegions,
                                 query = gapSummary) != 0
table(GapsBlacklisted)
gapSummary <- gapSummary[!GapsBlacklisted]
```

```{r}
### Dataframe of peakPresence
peakSummary.df <- as.data.frame(peakSummary@elementMetadata)
rownames(peakSummary.df)<- peakSummary@ranges@NAMES
colnames(peakSummary.df)<- unlist(lapply(strsplit(colnames(peakSummary.df),split ="_peaks"), function(x) x[[1]]))
```

###### Histogram on peak width

```{r, warning=FALSE}
ggplot(data.frame(peakSummary@ranges), aes(width)) + 
  geom_histogram(aes(y=..count..),binwidth = 10,fill="#A9DDD9")+
  scale_x_continuous(limits=c(0,200))+
  geom_text(aes(900, 200, label = paste(sum(data.frame(peakSummary@ranges)$width>1000),">",sep=""), 
                vjust = -0.5), col = "black")+
  theme_classic()
```

### Import Counts

```{r,warning=FALSE}
# count reads from bam files in peak regions
countsPerPeak <- peakCountTable(peaks = peakSummary, 
                                   path = bamFilePath, 
                                   pattern = ".final.bam$",
                                   )
#write peak counts into data frame
countsPerPeak.df <- as.data.frame(assay(countsPerPeak))
```

```{r,warning=FALSE}
# count reads from bam files in gap regions
countsPerGap <- peakCountTable(peaks = gapSummary, 
                               path = bamFilePath, 
                               pattern = ".final.bam$",
                               )

#write gap counts into data frame
countsPerGap.df <- as.data.frame(assay(countsPerGap))
```

### Export unfiltered counts

```{r}
# write.csv(countsPerPeak.df, "/home/caterina/data/analysis/counts_GEO.csv")
```

##### QC 

```{r}
# Count Statistics
totalPeakLength <- sum(as.numeric(width(peakSummary)))
totalPeakNumber <- length(peakSummary)
meanPeakLength <- round(totalPeakLength/totalPeakNumber,3)
totalPeakReads <- colSums(countsPerPeak.df)
meanReadsperPeak <- round(totalPeakReads/totalPeakNumber,3)
PeakReadsperBase <- totalPeakReads/totalPeakLength

# Gap statistics
totalGapLength <- sum(as.numeric(width(gapSummary)))
totalGapNumber <- length(gapSummary)
meanGapLength <- round(totalGapLength/totalGapNumber,3)
totalGapReads <- colSums(assay(countsPerGap))
meanReadsperGap <- round(totalGapReads/totalGapNumber,3)
GapReadsperBase <- totalGapReads/totalGapLength

summary.stats <- data.frame("totalPeakLength"=totalPeakLength,
                           "totalGapLength"=totalGapLength,
                            "totalPeakNumber"=totalPeakNumber,
                            "totalReads" = totalPeakReads+totalGapReads,
                            "totalPeakReads" = totalPeakReads,
                            "totalGapReads" = totalGapReads,
                            "meanReadsperPeak"= meanReadsperPeak,
                            "meanReadsperGap" = meanReadsperGap,
                            "PeakReadsperBase"=round(PeakReadsperBase,3),
                            "GapReadsperBase"= round(GapReadsperBase,3),
                            #"CoverageRatio" = round(PeakReadsperBase/GapReadsperBase,3))
                            "FRiP_score" = round(totalPeakReads/(totalPeakReads+totalGapReads),3))
idx <- match(rownames(summary.stats), sample_table$sample)
summary.stats$time <- sample_table$Time[idx]
summary.stats$group <- sample_table$Group[idx]
summary.stats$celltype <- sample_table$Celltype[idx]

summary.stats
```


```{r,fig.width=24,fig.height=6}
tmp <- melt(countsPerPeak.df)
idx <- match(tmp$variable,sample_table$ID)
tmp$ID <- sample_table$ID[idx]

p1 <- ggplot(tmp, aes(x= variable, y= value+1, fill= ID))+
  geom_boxplot() +
  scale_y_log10()+
  ylab("log10(Count+1)")+
  xlab("") +
  theme_classic()+
  theme(axis.text.x = element_text(size=8, angle = 45, hjust = 1),
        plot.title = element_text(size = 8, face = "bold"))

p2 <- ggplot(summary.stats, aes(x= rownames(summary.stats), y= totalPeakReads,
                                fill=rownames(summary.stats)))+
  geom_boxplot() +
  geom_jitter(aes(shape=rownames(summary.stats))) +
  ylab("Total Peak Counts")+
  xlab("") +
  theme_classic()+
  theme(axis.text.x = element_text(size=8, angle = 45, hjust = 1),
        plot.title = element_text(size = 8, face = "bold"))

# visualize the FRiP score
p3 <- ggplot(summary.stats, aes(x = rownames(summary.stats), y = FRiP_score, fill=rownames(summary.stats))) +
  geom_boxplot() +
  geom_jitter(aes(shape=rownames(summary.stats))) +
  ylab("Reads in Peaks/Reads in Gaps")+
  xlab("") +
  theme_classic()+
  theme(axis.text.x = element_text(size=8, angle = 45, hjust = 1),
        plot.title = element_text(size = 8, face = "bold"))

plot_grid(p1, p2, p3, ncol=2, rel_widths = c(2.3,1))
```

#### QC board

```{r, fig.width=14, fig.height=3.5}
# format TSS-ES
TSS_persample <- t(as.data.frame(TSS_df))
colnames(TSS_persample)[1] <- "TSS_ES"
TSS_persample <- as.data.frame(TSS_persample)
TSS_persample$ID <- rownames(TSS_persample)
reduced_st <- sample_table[, c(1, 2)]
TSS_persample <- merge(reduced_st, TSS_persample, by="ID")

# format FRiP score and log10(unique_reads)
FRiP_UR <- summary.stats[, c(4,11)]
FRiP_UR$log10UR <- log10(FRiP_UR$totalReads)
FRiP_UR$ID <- rownames(FRiP_UR)
FRiP_UR <- merge(reduced_st, FRiP_UR, by="ID")

# format mito%
mito_perc <- read.csv("/home/caterina/data/analysis/lorenzo/mito_counts/mito_percent.csv", header = FALSE)
colnames(mito_perc)[2] <- "mito%"
mito_perc$ID <- paste0("Sample_", mito_perc$V1)
mito_perc$V1 <- NULL
mito_perc <- merge(reduced_st, mito_perc, by="ID")

# plot board as one-col heatmap

# TSS_ES
my_breaks_TSS <- c(min(TSS_persample$TSS_ES), max(TSS_persample$TSS_ES))

TSS_hm <- ggplot(TSS_persample, aes(1, sample_name, fill = TSS_ES)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("#e5f9ec", "#008e2f"), breaks = my_breaks_TSS) +
  geom_text(aes(label = round(TSS_ES, 2))) +
  xlab("TSS_ES") +
  ylab("") +
  theme_bw(base_size = 13) +
  theme(plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.text=element_text(size=8),
    aspect.ratio = 5)

# FRiP
my_breaks_FRiP <- c(min(FRiP_UR$FRiP_score), max(FRiP_UR$FRiP_score))

FRiP_hm <- ggplot(FRiP_UR, aes(1, sample_name, fill = FRiP_score)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("#f6cef6", "#7e067e"), breaks = my_breaks_FRiP) +
  geom_text(aes(label = round(FRiP_score, 2))) +
  xlab("FRiP_score") +
  ylab("") +
  theme_bw(base_size = 13) +
  theme(plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.text=element_text(size=8),
    aspect.ratio = 5)

# log10UF
my_breaks_log10UR <- c(min(FRiP_UR$log10UR), max(FRiP_UR$log10UR))

log10UR_hm <- ggplot(FRiP_UR, aes(1, sample_name, fill = log10UR)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("#feefd8", "#e19e39"), breaks = my_breaks_log10UR) +
  geom_text(aes(label = round(log10UR, 2))) +
  xlab("log10UR") +
  ylab("") +
  theme_bw(base_size = 13) +
  theme(plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.text=element_text(size=8),
    aspect.ratio = 5)

# mito%
my_breaks_mito<- c(min(mito_perc$`mito%`), max(mito_perc$`mito%`))

mito_perc_hm <- ggplot(mito_perc, aes(1, sample_name, fill = `mito%`)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("#DE3E1C", "#f8d8d1"), breaks = my_breaks_mito) +
  geom_text(aes(label = round(`mito%`, 2))) +
  xlab("mito%") +
  ylab("") +
  theme_bw(base_size = 13) +
  theme(plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    legend.text=element_text(size=8),
    aspect.ratio = 5)

grid.arrange(TSS_hm, FRiP_hm, log10UR_hm, mito_perc_hm, ncol=4)
```

### Histogram of rowSums

```{r,fig.height=4,fig.width=8, warning=FALSE}
countStats <- data.frame(row.names=rownames(countsPerPeak.df),
                         mean=rowMeans(countsPerPeak.df),
                         sum=rowSums(countsPerPeak.df))

p4 <-ggplot(countStats, aes(mean)) + 
  geom_histogram(binwidth = 1, fill= "#A9DDD9")+
  scale_x_continuous(limits=c(0,50))+
  geom_text(aes(450, 1000, label = paste(sum(countStats$mean>500),">",sep=""), vjust = -0.5), col = "black")+
  theme_classic()

p5 <-ggplot(countStats, aes(sum)) + 
  geom_histogram(binwidth = 10, fill= "#A9DDD9")+
  scale_x_continuous(limits=c(0,200))+
  geom_text(aes(4950, 200, label = paste(sum(countStats$sum>5000),">",sep=""), vjust = -0.5), col = "black")+
  geom_vline(xintercept=10)+
  theme_classic()

plot_grid(p4, p5, labels=list("Mean","Sum"), 
          label_x = 0.5, label_y = 0.9, hjust = -0.5, vjust = -0.5)
```

### Filter Peaks

```{r}
peaks2keep <- rownames(countsPerPeak.df[rowMax(as.matrix(countsPerPeak.df))>10,])
length(peaks2keep)

# filter peaks
peakSummary.filt <- peakSummary[peaks2keep]
peakSummary.filt.df <- peakSummary.df[rownames(peakSummary.df) %in% peaks2keep,]

# filter counts
countsPerPeak.filt.df <- countsPerPeak.df[peaks2keep,]
```

### Peak Annotation

#### Genomic distribution

```{r, fig.height=16,fig.width=8}
peakSummary.filt$sum <- as.vector(rowSums(countsPerPeak.filt.df))
covplot(peakSummary.filt, weightCol="sum",title = "ATAC peaks over Chromosomes")
```

### Relative localization to nearest gene

```{r,warning=FALSE}
# complete peak set
peakAnnotation <- annotatePeak(peakSummary, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, level="gene")
peakAnnotation.df <- as.data.frame(as.GRanges(peakAnnotation))

# filtered peak set
peakAnnotation.filt <- annotatePeak(peakSummary.filt, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, level="gene")
peakAnnotation.filt.df <- as.data.frame(as.GRanges(peakAnnotation.filt))
```

##### PieChart

```{r}
plotAnnoPie(peakAnnotation.filt)
```

##### Histogram

```{r, fig.width=5, fig.height=4, warning=FALSE}
ggplot(data=peakAnnotation.filt.df, aes(x = peakAnnotation.filt.df$distanceToTSS)) + 
  geom_histogram(aes(y=..count..), colour="black", fill="white", binwidth = 50) +
  scale_x_continuous(limits=c(-2000,2000))+
  scale_y_continuous(limits=c(0,2000), oob = scales::squish)+
  labs(x="Distance to TSS (bp)", y="Peaks number")+
  theme_classic()+
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10))
```

# DESeq2 analysis of peak counts

#### DESeq2 object

```{r}
# produce DESeq data set
dds <- DESeqDataSetFromMatrix(countData = countsPerPeak.filt.df,
                              colData = sample_table,
                              design = ~ condition) # change the design
```

#### Calculate DESeq2 Statistics

```{r}
dds <- DESeq(dds,quiet = FALSE)
```

#### Variance Stabilization

```{r}
dds_rlog <- rlog(dds, blind = FALSE)
```

#### Principal Component Analysis

```{r, fig.height=3, fig.width=4}
plotPCA(pca_input = dds_rlog,
              ntop="all", 
              xPC=1, 
              yPC=2,
              point_size=4,
              title="",
              color="condition",
              anno_color=col_condition) +
  theme(aspect.ratio = 1)+
  theme(axis.text.x = element_text(colour = "black", size = 10), axis.text.y = element_text(colour = "black", size = 10))
```

## Annotated rlog normalized count table

```{r}
#write table of vst normalized counts
rlog_matrix <- as.matrix(assay(dds_rlog))
norm <- subset(rlog_matrix, rownames(rlog_matrix) %in% rownames(peakAnnotation.filt.df))

identical(rownames(norm), rownames(peakAnnotation.filt.df))

norm_anno <- cbind(norm, peakAnnotation.filt.df)
norm_anno$peakid <- rownames(norm_anno)
```

```{r}
# add symbol annotation

mart <- useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
getBM <- getBM(attributes = c("ensembl_gene_id", "mgi_symbol", "entrezgene_id"), bmHeader = T, mart = mart)
colnames(getBM)[3] <- "geneId"
colnames(getBM)[2] <- "symbol"
colnames(getBM)[1] <- "ensId"

#check out here

getBM_overlap <- subset(getBM, getBM$geneId %in% norm_anno$geneId)
getBM_overlap <- getBM_overlap[,c(2,3)]
getBM_overlap <- unique(getBM_overlap)
getBM_overlap <- getBM_overlap[!duplicated(getBM_overlap$geneId),]
norm_anno_symbol <- subset(norm_anno, norm_anno$geneId %in% getBM_overlap$geneId)
norm_anno_symbol <- merge(norm_anno_symbol, getBM_overlap, by = "geneId")
norm_anno <- norm_anno_symbol
```

# Define relevant comparisons

```{r}
comparison_table<-data.frame(comparison = c("LU50SCA","LU10BPA"),
                             control = c("BR50SCA","BR10BPA"))
```

# Perform Differential Expression Testing

```{r}
DEresults <- DEAnalysis(condition = "condition",
                        alpha=0.05 ,
                        lfcThreshold= 0,
                        sigFC = 2,
                        multiple_testing="IHW",
                        shrinkage = TRUE,
                        shrinkType="normal")
```

### Summary of DARs

```{r}
#all DARs
DEcounts <- NULL
for(i in 1:nrow(comparison_table)){
  tmp <- unlist(DEresults[[1+i]]@Number_DE_genes)
  DEcounts <- rbind(DEcounts, tmp)
}
rownames(DEcounts) <- names(DEresults)[-1]
DEcounts
```

### Upset plot

```{r, fig.height=3, fig.width=4}
DAR_up_50 <- uDARall_up(comparison = "LU50SCA_vs_BR50SCA")
DAR_up_BA <- uDARall_up(comparison = "LU10BPA_vs_BR10BPA")

intersect_DAR_up <- intersect(DAR_up_50, DAR_up_BA) #251

DAR_down_50 <- uDARall_down(comparison = "LU50SCA_vs_BR50SCA")
DAR_down_BA <- uDARall_down(comparison = "LU10BPA_vs_BR10BPA")

intersect_DAR_down <- intersect(DAR_down_50, DAR_down_BA) #411

combined_up_input <- list(DAR_up_50 = DAR_up_50,
                          DAR_up_BA = DAR_up_BA)
combined_down_input <- list(DAR_down_50 = DAR_down_50,
                          DAR_down_BA = DAR_down_BA)

upset(fromList(combined_up_input), nsets = 2, sets = c("DAR_up_50", "DAR_up_BA"), keep.order = T, order.by = "freq", mainbar.y.max = 800)

upset(fromList(combined_down_input), nsets = 2, sets = c("DAR_down_50", "DAR_down_BA"), keep.order = T, order.by = "freq", mainbar.y.max = 800)
```

## Heatmap of up and downregulated DARs mapping in promoters

```{r}
DAR_promo_up <- uDAR_promo_up(comparison = c("LU50SCA_vs_BR50SCA", "LU10BPA_vs_BR10BPA"))
DAR_promo_down <- uDAR_promo_down(comparison = c("LU50SCA_vs_BR50SCA", "LU10BPA_vs_BR10BPA"))

promo_union <- unique(c(DAR_promo_up, DAR_promo_down))

hm_normanno <- subset(norm_anno, norm_anno$peakid %in% promo_union)
rownames(hm_normanno) <- hm_normanno$peakid
hm_normanno_peaks <- hm_normanno[, 2:11]

hm_normanno_symbol <- hm_normanno[, c(2:11, 26)]
hm_normanno_symbol$var <- rowVars(hm_normanno_symbol[ ,1:10])
hm_normanno_symbol <- arrange(hm_normanno_symbol, desc(var)) 
hm_normanno_symbol <- hm_normanno_symbol[ !duplicated(hm_normanno_symbol$symbol), ]
rownames(hm_normanno_symbol) <- hm_normanno_symbol$symbol
hm_normanno_symbol$symbol <- NULL
hm_normanno_symbol$var <- NULL

hm_normanno_symbol <- scale(t(hm_normanno_symbol))
hm_normanno_symbol <- t(hm_normanno_symbol)
hm_normanno_symbol <- as.data.frame(hm_normanno_symbol)
```

```{r, fig.height=8, fig.width=4}
set.seed(1)

viridis(100)

HM_clusterized <- function(data, row_split = 5){
  
  col_fun <- colorRamp2(seq(-1.9, 1.9, length.out = 100), inferno(100))
  
  output <- list()
  output[["hm"]] <- Heatmap(as.matrix(data),
                col = col_fun, na_col = "black",border = T,
                show_row_names = FALSE,
                cluster_columns = FALSE,
                use_raster = F, 
                row_split = row_split)
  clusters <- row_order(output$hm)
  for(i in 1:length(clusters)){
    clusters[[i]] <- rownames(data)[clusters[[i]]]
  }
  
  output[["clusters"]] <- clusters
  output[["cluster_lengths"]] <- as.list(as.character(lengths(output$clusters)))
  
  return(output)
}

HM_clust <- HM_clusterized(hm_normanno_symbol, row_split = 2)
HM_clust$hm
```

### Define universe and gene sets for subsequent GSEA analyses

```{r}
# define universe
universe <- unique(as.character(norm_anno$symbol))
# change symbols to ENTREZ IDs (necessary for ClusterProfiler)
universe_Entrez <- bitr(universe, 
                        fromType="SYMBOL", 
                        toType="ENTREZID", 
                        OrgDb="org.Mm.eg.db")$ENTREZID
```

### HM modules enrichment

```{r}
clusters <- HM_clust$clusters
  
collector_GO <- list()
  
  for (n in 1:length(clusters)) {
    
    gene_list <- clusters[[n]]
    
    entrez_list <- bitr(gene_list, fromType = "SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")$ENTREZID
    
    tmp <- enrichGO(gene = entrez_list,
          universe = universe_Entrez,
          OrgDb = "org.Mm.eg.db",
          ont = "BP",
          pAdjustMethod = "bonferroni",
          pvalueCutoff  = 1,
          qvalueCutoff  = 1,
          readable      = T)
    
    collector_GO[[n]] <- as.data.frame(tmp)
  }
```

## Dotplot

```{r}
GOterms <- c("GO:0030324", "GO:0060425", "GO:0048286", "GO:0060441", "GO:0060479", "GO:0060487", "GO:0050808", "GO:0050804", "GO:0048667", "GO:0007420", "GO:0030902","GO:0021533")

GSEA_1 <- collector_GO[[1]]
GO_input_1 <- subset(GSEA_1, ID %in% GOterms)
GO_input_1$Cluster <- "cluster1"
GO_input_1 <- GO_input_1[match(GOterms, GO_input_1$ID),]
GO_input_1 <- na.omit(GO_input_1)

GSEA_2 <- collector_GO[[2]]
GO_input_2 <- subset(GSEA_2, ID %in% GOterms)
GO_input_2$Cluster <- "cluster2"
GO_input_2 <- GO_input_2[match(GOterms, GO_input_2$ID),]
GO_input_2 <- na.omit(GO_input_2)

ClusterUPDOWN <- rbind(GO_input_1, GO_input_2)
ClusterUPDOWN$Cluster <- factor(ClusterUPDOWN$Cluster, levels = c("cluster1", "cluster2"))
ClusterUPDOWN$Numbering <- 1:nrow(ClusterUPDOWN) 
ClusterUPDOWN$Description <- fct_rev(factor(ClusterUPDOWN$Description, levels = (ClusterUPDOWN[c(1:12) ,2])))

dotplotGSEA_MC(ClusterUPDOWN,
            font.size = 12,
            title.size=12,
            title.width = 40)
```

# Genome browser tracks

```{r, fig.height=10, fig.width=8}
# Rundc3a
# afrom <- 102265280
# ato <- 102465281
# chr 11

# Napsa
# afrom <- 44483875
# ato <- 44675415
# chr 7

# Sftpb
afrom <- 72290893
ato <- 72408090
# chr 6

gtrack <- GenomeAxisTrack()

options(ucscChromosomeNames=FALSE)
rtrack <- UcscTrack(genome = "mm10", chromosome = "chr6",
                      track = "All GENCODE VM16", from = afrom, to = ato,
                      trackType = "GeneRegionTrack", 
                      rstarts = "exonStarts", rends = "exonEnds", 
                      gene = "name",  symbol = "name2", 
                      transcript = "name", strand = "strand",
                      fill = "#8282d2", 
                      name = "Other RefSeq", showID = T, geneSymbol = T)

bw_lung_encode <- "/home/caterina/data/analysis/ENCODE_bw/ENCFF435PTI.bigWig" #pval reference
bw_brain_encode <- "/home/caterina/data/analysis/ENCODE_bw/ENCFF561KNB_brainpval.bigWig" #pval reference

# bw_lung50_1 <- "/home/caterina/data/alignment/2022-10-26_lung_brain_nobulk/output/mapped/Sample_15838.final.bw"
# bw_lung50_2 <- "/home/caterina/data/alignment/2022-10-26_lung_brain_nobulk/output/mapped/Sample_15839.final.bw"
# bw_lung50_3 <- "/home/caterina/data/alignment/2022-10-26_lung_brain_nobulk/output/mapped/Sample_15840.final.bw"

#Substitute track with RPKM norm tracks

bw_lung50_1 <- "/home/caterina/data/analysis/lorenzo/norm_bw/Sample_15838.final.bw"
bw_lung50_2 <- "/home/caterina/data/analysis/lorenzo/norm_bw/Sample_15839.final.bw"
bw_lung50_3 <- "/home/caterina/data/analysis/lorenzo/norm_bw/Sample_15840.final.bw"

# bw_lung200_1 <- "/home/caterina/data/alignment/2022-10-26_lung_brain_nobulk/output/mapped/Sample_15841.final.bw"
# bw_lung200_2 <- "/home/caterina/data/alignment/2022-10-26_lung_brain_nobulk/output/mapped/Sample_15842.final.bw"
# bw_lung200_3 <- "/home/caterina/data/alignment/2022-10-26_lung_brain_nobulk/output/mapped/Sample_15843.final.bw"

bw_lung200_1 <- "/home/caterina/data/analysis/lorenzo/norm_bw/Sample_15841.final.bw"
bw_lung200_2 <- "/home/caterina/data/analysis/lorenzo/norm_bw/Sample_15842.final.bw"
bw_lung200_3 <- "/home/caterina/data/analysis/lorenzo/norm_bw/Sample_15843.final.bw"

# bw_brain50_1 <- "/home/caterina/data/alignment/2022-10-26_lung_brain_nobulk/output/mapped/Sample_15847.final.bw"
# bw_brain50_2 <- "/home/caterina/data/alignment/2022-10-26_lung_brain_nobulk/output/mapped/Sample_15848.final.bw"

bw_brain50_1 <- "/home/caterina/data/analysis/lorenzo/norm_bw/Sample_15847.final.bw"
bw_brain50_2 <- "/home/caterina/data/analysis/lorenzo/norm_bw/Sample_15848.final.bw"

# bw_brain200_1 <- "/home/caterina/data/alignment/2022-10-26_lung_brain_nobulk/output/mapped/Sample_15850.final.bw"
# bw_brain200_2 <- "/home/caterina/data/alignment/2022-10-26_lung_brain_nobulk/output/mapped/Sample_15851.final.bw"

bw_brain200_1 <- "/home/caterina/data/analysis/lorenzo/norm_bw/Sample_15850.final.bw"
bw_brain200_2 <- "/home/caterina/data/analysis/lorenzo/norm_bw/Sample_15851.final.bw"

# Lung
bwtrack_l_encode <- DataTrack(bw_lung_encode, isPaired = T, name = "l_encode", type  = 'hist', fill = "grey")

bwtrack_l_50_1 <- DataTrack(bw_lung50_1, isPaired = T, groups = factor("sample 1", levels = c("sample 1", "sample 2", "sample 3")), name = "l_50", type  = 'hist')
bwtrack_l_50_2 <- DataTrack(bw_lung50_2, isPaired = T, groups = factor("sample 2", levels = c("sample 1", "sample 2", "sample 3")), name = "l_50", type  = 'hist')
bwtrack_l_50_3 <- DataTrack(bw_lung50_3, isPaired = T, groups = factor("sample 3", levels = c("sample 1", "sample 2", "sample 3")), name = "l_50", type  = 'hist')

displayPars(bwtrack_l_50_1) <- list(alpha.title = 1, alpha = 1, col = "#ff1919", name = "l_50", col.histogram = "NA")
displayPars(bwtrack_l_50_2) <- list(alpha.title = 1, alpha = 1, col = "#ff4c4c", name = "l_50", col.histogram = "NA")
displayPars(bwtrack_l_50_3) <- list(alpha.title = 1, alpha = 1, col = "#ff7f7f", name = "l_50", col.histogram = "NA")

ot_l_50_bw <- OverlayTrack(trackList=list(bwtrack_l_50_1, bwtrack_l_50_2, bwtrack_l_50_3)) 

bwtrack_l_200_1 <- DataTrack(bw_lung200_1, isPaired = T, groups = factor("sample 1", levels = c("sample 1", "sample 2", "sample 3")), name = "l_200", type  = 'hist')
bwtrack_l_200_2 <- DataTrack(bw_lung200_2, isPaired = T, groups = factor("sample 2", levels = c("sample 1", "sample 2", "sample 3")), name = "l_200", type  = 'hist')
bwtrack_l_200_3 <- DataTrack(bw_lung200_3, isPaired = T, groups = factor("sample 3", levels = c("sample 1", "sample 2", "sample 3")), name = "l_200", type  = 'hist')

displayPars(bwtrack_l_200_1) <- list(alpha.title = 1, alpha = 1, col = "#ff1919", name = "l_200", col.histogram = "NA")
displayPars(bwtrack_l_200_2) <- list(alpha.title = 1, alpha = 1, col = "#ff4c4c", name = "l_200", col.histogram = "NA")
displayPars(bwtrack_l_200_3) <- list(alpha.title = 1, alpha = 1, col = "#ff7f7f", name = "l_200", col.histogram = "NA")

ot_l_200_bw <- OverlayTrack(trackList=list(bwtrack_l_200_1, bwtrack_l_200_2, bwtrack_l_200_3)) 

# Brain
bwtrack_b_encode <- DataTrack(bw_brain_encode, isPaired = T, name = "b_encode", type  = 'hist', fill = "grey")

bwtrack_b_50_1 <- DataTrack(bw_brain50_1, isPaired = T, groups = factor("sample 1", levels = c("sample 1", "sample 2")), name = "b_50", type  = 'hist')
bwtrack_b_50_2 <- DataTrack(bw_brain50_2, isPaired = T, groups = factor("sample 2", levels = c("sample 1", "sample 2")),  name = "b_50", type  = 'hist')

displayPars(bwtrack_b_50_1) <- list(alpha.title = 1, alpha = 1, col = "#b266b2", name = "b_50", col.histogram = "NA")
displayPars(bwtrack_b_50_2) <- list(alpha.title = 1, alpha = 1, col = "#cc99cc", name = "b_50", col.histogram = "NA")

ot_b_50_bw <- OverlayTrack(trackList=list(bwtrack_b_50_1, bwtrack_b_50_2)) 

bwtrack_b_200_1 <- DataTrack(bw_brain200_1, isPaired = T, groups = factor("sample 1", levels = c("sample 1", "sample 2")), name = "b_200", type  = 'hist')
bwtrack_b_200_2 <- DataTrack(bw_brain200_2, isPaired = T, groups = factor("sample 2", levels = c("sample 1", "sample 2")), name = "b_200", type  = 'hist')

displayPars(bwtrack_b_200_1) <- list(alpha.title = 1, alpha = 1, col = "#b266b2", name = "b_200", col.histogram = "NA")
displayPars(bwtrack_b_200_2) <- list(alpha.title = 1, alpha = 1, col = "#cc99cc", name = "b_200", col.histogram = "NA")

ot_b_200_bw <- OverlayTrack(trackList=list(bwtrack_b_200_1, bwtrack_b_200_2)) 

options(ucscChromosomeNames=FALSE)

plotTracks(c(bwtrack_l_encode, ot_l_50_bw, ot_l_200_bw, bwtrack_b_encode, ot_b_50_bw, ot_b_200_bw, gtrack, rtrack), from = afrom, to = ato, type = "hist", window = -1, chromosome = "chr6", transcriptAnnotation = "symbol", sizes=c(1,1,1,1,1,1,1,1), legend = FALSE, col.title = "black", col.axis = "black", cex.axis = 0.5) 
```

# Session info

```{r}
sessionInfo <- sessionInfo()
```


